{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7106dcdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from collections import OrderedDict, defaultdict\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio.SeqUtils.IsoelectricPoint import IsoelectricPoint as IP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "701a2d9d",
   "metadata": {},
   "source": [
    "**Data:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "752be6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_dir = 'metadata'\n",
    "upstreams_dir = 'upstream_search'\n",
    "res_dir = 'results'\n",
    "\n",
    "hosts_file = os.path.join(meta_dir, 'seq_id_to_host.tsv')\n",
    "faa_file = os.path.join(upstreams_dir, 'upstream.faa')\n",
    "gff_file = os.path.join(res_dir, 'upstreams_clusters_domains.gff')\n",
    "res_file = os.path.join(res_dir, 'upstream_proteins_clu_wide_seq_sorted_prokka_domains.tsv')\n",
    "\n",
    "output_file = os.path.join(res_dir, 'upstream_proteins_clu_wide_seq_sorted_prokka_domains_pi_length_host.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b8bdad",
   "metadata": {},
   "source": [
    "**Read info prior information about clusters**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8af73124",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clu_parent</th>\n",
       "      <th>count</th>\n",
       "      <th>products</th>\n",
       "      <th>sequence</th>\n",
       "      <th>phrOG</th>\n",
       "      <th>phrOG_annotation</th>\n",
       "      <th>phrOG_category</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "      <th>cdd</th>\n",
       "      <th>cdd_name</th>\n",
       "      <th>tigr</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>cluster_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_001470995.1_00023</td>\n",
       "      <td>354</td>\n",
       "      <td>{T7 RNA polymerase: 354}</td>\n",
       "      <td>MEYNPLTELASLYGEDLAAEQLRLEAEAYSLGEKRFMEAMEFKAET...</td>\n",
       "      <td>phrog_20599,phrog_414</td>\n",
       "      <td>RNA polymerase</td>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cl42978,cl34907,cl44477,cl20638</td>\n",
       "      <td>PHA00452,RPO41,RNA_pol,RPOL_N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_025085155.1_00014</td>\n",
       "      <td>190</td>\n",
       "      <td>{Protein kinase 0.7: 183, hypothetical protein...</td>\n",
       "      <td>MYAHFQSLLKAIRELPINELDKRQPMLVDLLAQIVNHETQDGAITA...</td>\n",
       "      <td>phrog_18305,phrog_2828,phrog_15752</td>\n",
       "      <td>kinase,serine-threonine kinase,protein kinase;...</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cl20113</td>\n",
       "      <td>PHA00451</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_001041815.1_00006</td>\n",
       "      <td>150</td>\n",
       "      <td>{Overcome classical restriction gp0.3: 102, hy...</td>\n",
       "      <td>MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...</td>\n",
       "      <td>phrog_2568</td>\n",
       "      <td>ocr-like anti-restriction</td>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_023716895.1_00035</td>\n",
       "      <td>149</td>\n",
       "      <td>{hypothetical protein: 149}</td>\n",
       "      <td>METSCRLPHGAGRSCTGMDGDAMMFTAFIVFMYLLIVLYFLKDFRK...</td>\n",
       "      <td>phrog_2385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCA_009672325.1_00007</td>\n",
       "      <td>144</td>\n",
       "      <td>{hypothetical protein: 144}</td>\n",
       "      <td>MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>681</th>\n",
       "      <td>GCA_005892785.1_00005</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MFKHEVFISDNREAAMIAEMFKGHVTAYNIGDDEEAFCRYLC</td>\n",
       "      <td>phrog_6403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>GCA_005892125.1_00006</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MKYILIAMFSGFIGYAGASYDNYRSKQLAAEKWDKTYFACMKQGNL...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>GCA_005280775.1_00002</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MVTYYVYVNQMEYFEADTLEEAIRLCDSVPGSHVEDSVLDEVIYANR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>GCA_002922415.1_00057</td>\n",
       "      <td>1</td>\n",
       "      <td>{Tail spike protein: 1}</td>\n",
       "      <td>VNSSVGVGSVVVKDSYIYYIFGGENHFNPMTYGDNKDKDPFKGHGH...</td>\n",
       "      <td>phrog_38424,phrog_17635,phrog_14895</td>\n",
       "      <td>tail fiber protein,tail spike protein</td>\n",
       "      <td>unknown function,tail</td>\n",
       "      <td>End_beta_propel,End_tail_spike,Peptidase_S74</td>\n",
       "      <td>PF12217.11,PF12219.11,PF13884.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cl13629,cl16452,cl28659</td>\n",
       "      <td>End_tail_spike,Peptidase_S74_CIMCD,End_beta_pr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>GCA_002921805.1_00012</td>\n",
       "      <td>1</td>\n",
       "      <td>{T7 RNA polymerase: 1}</td>\n",
       "      <td>VELEAQMRPAAILKFKKNMLSLVNSDNAISTPPVYSLLINSLKTLE...</td>\n",
       "      <td>phrog_2051,phrog_20599,phrog_36028</td>\n",
       "      <td>RNA polymerase</td>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cl42978,cl34907,cl44477</td>\n",
       "      <td>PHA00452,RPO41,RNA_pol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>685 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                clu_parent  count   \n",
       "1    GCF_001470995.1_00023    354  \\\n",
       "2    GCA_025085155.1_00014    190   \n",
       "3    GCF_001041815.1_00006    150   \n",
       "4    GCA_023716895.1_00035    149   \n",
       "5    GCA_009672325.1_00007    144   \n",
       "..                     ...    ...   \n",
       "681  GCA_005892785.1_00005      1   \n",
       "682  GCA_005892125.1_00006      1   \n",
       "683  GCA_005280775.1_00002      1   \n",
       "684  GCA_002922415.1_00057      1   \n",
       "685  GCA_002921805.1_00012      1   \n",
       "\n",
       "                                              products   \n",
       "1                             {T7 RNA polymerase: 354}  \\\n",
       "2    {Protein kinase 0.7: 183, hypothetical protein...   \n",
       "3    {Overcome classical restriction gp0.3: 102, hy...   \n",
       "4                          {hypothetical protein: 149}   \n",
       "5                          {hypothetical protein: 144}   \n",
       "..                                                 ...   \n",
       "681                          {hypothetical protein: 1}   \n",
       "682                          {hypothetical protein: 1}   \n",
       "683                          {hypothetical protein: 1}   \n",
       "684                            {Tail spike protein: 1}   \n",
       "685                             {T7 RNA polymerase: 1}   \n",
       "\n",
       "                                              sequence   \n",
       "1    MEYNPLTELASLYGEDLAAEQLRLEAEAYSLGEKRFMEAMEFKAET...  \\\n",
       "2    MYAHFQSLLKAIRELPINELDKRQPMLVDLLAQIVNHETQDGAITA...   \n",
       "3    MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...   \n",
       "4    METSCRLPHGAGRSCTGMDGDAMMFTAFIVFMYLLIVLYFLKDFRK...   \n",
       "5    MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...   \n",
       "..                                                 ...   \n",
       "681         MFKHEVFISDNREAAMIAEMFKGHVTAYNIGDDEEAFCRYLC   \n",
       "682  MKYILIAMFSGFIGYAGASYDNYRSKQLAAEKWDKTYFACMKQGNL...   \n",
       "683    MVTYYVYVNQMEYFEADTLEEAIRLCDSVPGSHVEDSVLDEVIYANR   \n",
       "684  VNSSVGVGSVVVKDSYIYYIFGGENHFNPMTYGDNKDKDPFKGHGH...   \n",
       "685  VELEAQMRPAAILKFKKNMLSLVNSDNAISTPPVYSLLINSLKTLE...   \n",
       "\n",
       "                                   phrOG   \n",
       "1                  phrog_20599,phrog_414  \\\n",
       "2     phrog_18305,phrog_2828,phrog_15752   \n",
       "3                             phrog_2568   \n",
       "4                             phrog_2385   \n",
       "5                             phrog_2226   \n",
       "..                                   ...   \n",
       "681                           phrog_6403   \n",
       "682                                  NaN   \n",
       "683                                  NaN   \n",
       "684  phrog_38424,phrog_17635,phrog_14895   \n",
       "685   phrog_2051,phrog_20599,phrog_36028   \n",
       "\n",
       "                                      phrOG_annotation   \n",
       "1                                       RNA polymerase  \\\n",
       "2    kinase,serine-threonine kinase,protein kinase;...   \n",
       "3                            ocr-like anti-restriction   \n",
       "4                                                  NaN   \n",
       "5                      SAM-dependent methyltransferase   \n",
       "..                                                 ...   \n",
       "681                                                NaN   \n",
       "682                                                NaN   \n",
       "683                                                NaN   \n",
       "684              tail fiber protein,tail spike protein   \n",
       "685                                     RNA polymerase   \n",
       "\n",
       "                                        phrOG_category   \n",
       "1                   DNA, RNA and nucleotide metabolism  \\\n",
       "2                                                other   \n",
       "3    moron, auxiliary metabolic gene and host takeover   \n",
       "4                                     unknown function   \n",
       "5                                                other   \n",
       "..                                                 ...   \n",
       "681                                   unknown function   \n",
       "682                                                NaN   \n",
       "683                                                NaN   \n",
       "684                              unknown function,tail   \n",
       "685                 DNA, RNA and nucleotide metabolism   \n",
       "\n",
       "                                  pfam_annotation   \n",
       "1                                  RNA_pol,RPOL_N  \\\n",
       "2                                             NaN   \n",
       "3                                             ocr   \n",
       "4                                             NaN   \n",
       "5                                             NaN   \n",
       "..                                            ...   \n",
       "681                                           NaN   \n",
       "682                                           NaN   \n",
       "683                                           NaN   \n",
       "684  End_beta_propel,End_tail_spike,Peptidase_S74   \n",
       "685                                RNA_pol,RPOL_N   \n",
       "\n",
       "                                pfam conj_pl_domains   \n",
       "1               PF00940.22,PF14700.9             NaN  \\\n",
       "2                                NaN             NaN   \n",
       "3                         PF08684.13             NaN   \n",
       "4                                NaN             NaN   \n",
       "5                                NaN             NaN   \n",
       "..                               ...             ...   \n",
       "681                              NaN             NaN   \n",
       "682                              NaN             NaN   \n",
       "683                              NaN             NaN   \n",
       "684  PF12217.11,PF12219.11,PF13884.9             NaN   \n",
       "685             PF00940.22,PF14700.9             NaN   \n",
       "\n",
       "                                 cdd   \n",
       "1    cl42978,cl34907,cl44477,cl20638  \\\n",
       "2                            cl20113   \n",
       "3                                NaN   \n",
       "4                                NaN   \n",
       "5                                NaN   \n",
       "..                               ...   \n",
       "681                              NaN   \n",
       "682                              NaN   \n",
       "683                              NaN   \n",
       "684          cl13629,cl16452,cl28659   \n",
       "685          cl42978,cl34907,cl44477   \n",
       "\n",
       "                                              cdd_name tigr tigrfam_name   \n",
       "1                        PHA00452,RPO41,RNA_pol,RPOL_N  NaN          NaN  \\\n",
       "2                                             PHA00451  NaN          NaN   \n",
       "3                                                  NaN  NaN          NaN   \n",
       "4                                                  NaN  NaN          NaN   \n",
       "5                                                  NaN  NaN          NaN   \n",
       "..                                                 ...  ...          ...   \n",
       "681                                                NaN  NaN          NaN   \n",
       "682                                                NaN  NaN          NaN   \n",
       "683                                                NaN  NaN          NaN   \n",
       "684  End_tail_spike,Peptidase_S74_CIMCD,End_beta_pr...  NaN          NaN   \n",
       "685                             PHA00452,RPO41,RNA_pol  NaN          NaN   \n",
       "\n",
       "     cluster_num  \n",
       "1              1  \n",
       "2              2  \n",
       "3              3  \n",
       "4              4  \n",
       "5              5  \n",
       "..           ...  \n",
       "681          681  \n",
       "682          682  \n",
       "683          683  \n",
       "684          684  \n",
       "685          685  \n",
       "\n",
       "[685 rows x 15 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_table = pd.read_csv(res_file, sep='\\t')\n",
    "res_table['cluster_num'] = res_table.index\n",
    "res_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bf91702",
   "metadata": {},
   "source": [
    "**Read information about hosts:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4e4bb6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "hosts_data = pd.read_csv(hosts_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "49c81a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "hosts_data.describe()\n",
    "hosts_data['is_Escherichia'] = hosts_data['order'] == 'Escherichia'\n",
    "hosts_data['is_Enterobacterales'] = hosts_data['order'] == 'Enterobacterales'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98b2c437",
   "metadata": {},
   "source": [
    "**Read GFF with information about domains, clusters and nuccore ids.**  \n",
    "Take only nessesary columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9086f277",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames = ['seq_id', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attribute']\n",
    "gff_data = pd.read_csv(gff_file, sep='\\t', skiprows=1, names=colnames)\n",
    "gff_data = gff_data.drop(columns=['source', 'type', 'score', 'phase'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdf873a7",
   "metadata": {},
   "source": [
    "Create a dict from attribute string:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b06998f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>attribute</th>\n",
       "      <th>attribute_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>901</td>\n",
       "      <td>1254</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=GCA_002600005.1_00001;file_id=upstreams_wit...</td>\n",
       "      <td>{'ID': 'GCA_002600005.1_00001', 'file_id': 'up...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1254</td>\n",
       "      <td>1409</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=GCA_002600005.1_00002;file_id=upstreams_wit...</td>\n",
       "      <td>{'ID': 'GCA_002600005.1_00002', 'file_id': 'up...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1478</td>\n",
       "      <td>1615</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=GCA_002600005.1_00003;file_id=upstreams_wit...</td>\n",
       "      <td>{'ID': 'GCA_002600005.1_00003', 'file_id': 'up...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1612</td>\n",
       "      <td>2256</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=GCA_002600005.1_00004;file_id=upstreams_wit...</td>\n",
       "      <td>{'ID': 'GCA_002600005.1_00004', 'file_id': 'up...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>2327</td>\n",
       "      <td>4978</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=GCA_002600005.1_00005;Name=1;file_id=upstre...</td>\n",
       "      <td>{'ID': 'GCA_002600005.1_00005', 'Name': '1', '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id  start   end strand   \n",
       "0  AY264775.1    901  1254      +  \\\n",
       "1  AY264775.1   1254  1409      +   \n",
       "2  AY264775.1   1478  1615      +   \n",
       "3  AY264775.1   1612  2256      +   \n",
       "4  AY264775.1   2327  4978      +   \n",
       "\n",
       "                                           attribute   \n",
       "0  ID=GCA_002600005.1_00001;file_id=upstreams_wit...  \\\n",
       "1  ID=GCA_002600005.1_00002;file_id=upstreams_wit...   \n",
       "2  ID=GCA_002600005.1_00003;file_id=upstreams_wit...   \n",
       "3  ID=GCA_002600005.1_00004;file_id=upstreams_wit...   \n",
       "4  ID=GCA_002600005.1_00005;Name=1;file_id=upstre...   \n",
       "\n",
       "                                      attribute_dict  \n",
       "0  {'ID': 'GCA_002600005.1_00001', 'file_id': 'up...  \n",
       "1  {'ID': 'GCA_002600005.1_00002', 'file_id': 'up...  \n",
       "2  {'ID': 'GCA_002600005.1_00003', 'file_id': 'up...  \n",
       "3  {'ID': 'GCA_002600005.1_00004', 'file_id': 'up...  \n",
       "4  {'ID': 'GCA_002600005.1_00005', 'Name': '1', '...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_data['attribute'] = gff_data['attribute'].apply(lambda x: x.replace('=;', '=NaN;'))  # fix a data artefact\n",
    "gff_data['attribute_dict'] = gff_data['attribute'].apply(lambda x: {i.split('=')[0]: \n",
    "                                                                    i.split('=')[1] for i in x.split(';')})\n",
    "gff_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "698aded2",
   "metadata": {},
   "source": [
    "Create separate columns from attribute dict:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23bf4ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_attribute = pd.json_normalize(gff_data.attribute_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ae4a9c50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>file_id</th>\n",
       "      <th>inference</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>cluster_main_prod</th>\n",
       "      <th>phrOG</th>\n",
       "      <th>phrOG_annotation</th>\n",
       "      <th>phrOG_category</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>e_c_number</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdd</th>\n",
       "      <th>cdd_name</th>\n",
       "      <th>tigr</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>Overcome classical restriction gp0.3</td>\n",
       "      <td>2</td>\n",
       "      <td>Overcome classical restriction gp0.3</td>\n",
       "      <td>phrog_2568</td>\n",
       "      <td>ocr-like anti-restriction</td>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>Gene 0.4 protein</td>\n",
       "      <td>12</td>\n",
       "      <td>Gene 0.4 protein</td>\n",
       "      <td>phrog_25044,phrog_7202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>18</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>phrog_16572</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>Protein kinase 0.7</td>\n",
       "      <td>134</td>\n",
       "      <td>protein kinase</td>\n",
       "      <td>phrog_8393,phrog_29241,phrog_2828,phrog_15752</td>\n",
       "      <td>serine-threonine kinase,protein kinase, inhibi...</td>\n",
       "      <td>unknown function,other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.11.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>T7 RNA polymerase</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA polymerase</td>\n",
       "      <td>phrog_2051,phrog_2166,phrog_414</td>\n",
       "      <td>RNA polymerase,N4-like RNA polymerase</td>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>2.7.7.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID                  file_id   \n",
       "0  GCA_002600005.1_00001  upstreams_with_clusters  \\\n",
       "1  GCA_002600005.1_00002  upstreams_with_clusters   \n",
       "2  GCA_002600005.1_00003  upstreams_with_clusters   \n",
       "3  GCA_002600005.1_00004  upstreams_with_clusters   \n",
       "4  GCA_002600005.1_00005  upstreams_with_clusters   \n",
       "\n",
       "                                           inference              locus_tag   \n",
       "0  ab initio prediction:Prodigal:2.6,similar to A...  GCA_002600005.1_00001  \\\n",
       "1  ab initio prediction:Prodigal:2.6,similar to A...  GCA_002600005.1_00002   \n",
       "2                  ab initio prediction:Prodigal:2.6  GCA_002600005.1_00003   \n",
       "3  ab initio prediction:Prodigal:2.6,similar to A...  GCA_002600005.1_00004   \n",
       "4  ab initio prediction:Prodigal:2.6,similar to A...  GCA_002600005.1_00005   \n",
       "\n",
       "                                product cluster_num   \n",
       "0  Overcome classical restriction gp0.3           2  \\\n",
       "1                      Gene 0.4 protein          12   \n",
       "2                  hypothetical protein          18   \n",
       "3                    Protein kinase 0.7         134   \n",
       "4                     T7 RNA polymerase           0   \n",
       "\n",
       "                      cluster_main_prod   \n",
       "0  Overcome classical restriction gp0.3  \\\n",
       "1                      Gene 0.4 protein   \n",
       "2                  hypothetical protein   \n",
       "3                        protein kinase   \n",
       "4                        RNA polymerase   \n",
       "\n",
       "                                           phrOG   \n",
       "0                                     phrog_2568  \\\n",
       "1                         phrog_25044,phrog_7202   \n",
       "2                                    phrog_16572   \n",
       "3  phrog_8393,phrog_29241,phrog_2828,phrog_15752   \n",
       "4                phrog_2051,phrog_2166,phrog_414   \n",
       "\n",
       "                                    phrOG_annotation   \n",
       "0                          ocr-like anti-restriction  \\\n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  serine-threonine kinase,protein kinase, inhibi...   \n",
       "4              RNA polymerase,N4-like RNA polymerase   \n",
       "\n",
       "                                      phrOG_category pfam_annotation   \n",
       "0  moron, auxiliary metabolic gene and host takeover             ocr  \\\n",
       "1                                   unknown function             NaN   \n",
       "2                                   unknown function             NaN   \n",
       "3                             unknown function,other             NaN   \n",
       "4                 DNA, RNA and nucleotide metabolism  RNA_pol,RPOL_N   \n",
       "\n",
       "                   pfam e_c_number Name gene  cdd cdd_name tigr tigrfam_name   \n",
       "0            PF08684.13        NaN  NaN  NaN  NaN      NaN  NaN          NaN  \\\n",
       "1                   NaN        NaN  NaN  NaN  NaN      NaN  NaN          NaN   \n",
       "2                   NaN        NaN  NaN  NaN  NaN      NaN  NaN          NaN   \n",
       "3                   NaN   2.7.11.1  NaN  NaN  NaN      NaN  NaN          NaN   \n",
       "4  PF00940.22,PF14700.9    2.7.7.6    1    1  NaN      NaN  NaN          NaN   \n",
       "\n",
       "  conj_pl_domains  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2             NaN  \n",
       "3             NaN  \n",
       "4             NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_attribute.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87655c2c",
   "metadata": {},
   "source": [
    "\"Append\" normalized attribute to gff dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bb5c9ba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_data = pd.concat([gff_data, norm_attribute], axis=1)\n",
    "# remove temp columns:\n",
    "gff_data = gff_data.drop(columns=['attribute', 'attribute_dict'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "866fa56d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>file_id</th>\n",
       "      <th>inference</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>...</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>e_c_number</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdd</th>\n",
       "      <th>cdd_name</th>\n",
       "      <th>tigr</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>901</td>\n",
       "      <td>1254</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>Overcome classical restriction gp0.3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1254</td>\n",
       "      <td>1409</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>Gene 0.4 protein</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1478</td>\n",
       "      <td>1615</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1612</td>\n",
       "      <td>2256</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>Protein kinase 0.7</td>\n",
       "      <td>134</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.11.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>2327</td>\n",
       "      <td>4978</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>T7 RNA polymerase</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>2.7.7.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id  start   end strand                     ID   \n",
       "0  AY264775.1    901  1254      +  GCA_002600005.1_00001  \\\n",
       "1  AY264775.1   1254  1409      +  GCA_002600005.1_00002   \n",
       "2  AY264775.1   1478  1615      +  GCA_002600005.1_00003   \n",
       "3  AY264775.1   1612  2256      +  GCA_002600005.1_00004   \n",
       "4  AY264775.1   2327  4978      +  GCA_002600005.1_00005   \n",
       "\n",
       "                   file_id                                          inference   \n",
       "0  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...  \\\n",
       "1  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "2  upstreams_with_clusters                  ab initio prediction:Prodigal:2.6   \n",
       "3  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "4  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "\n",
       "               locus_tag                               product cluster_num   \n",
       "0  GCA_002600005.1_00001  Overcome classical restriction gp0.3           2  \\\n",
       "1  GCA_002600005.1_00002                      Gene 0.4 protein          12   \n",
       "2  GCA_002600005.1_00003                  hypothetical protein          18   \n",
       "3  GCA_002600005.1_00004                    Protein kinase 0.7         134   \n",
       "4  GCA_002600005.1_00005                     T7 RNA polymerase           0   \n",
       "\n",
       "   ... pfam_annotation                  pfam e_c_number Name gene  cdd   \n",
       "0  ...             ocr            PF08684.13        NaN  NaN  NaN  NaN  \\\n",
       "1  ...             NaN                   NaN        NaN  NaN  NaN  NaN   \n",
       "2  ...             NaN                   NaN        NaN  NaN  NaN  NaN   \n",
       "3  ...             NaN                   NaN   2.7.11.1  NaN  NaN  NaN   \n",
       "4  ...  RNA_pol,RPOL_N  PF00940.22,PF14700.9    2.7.7.6    1    1  NaN   \n",
       "\n",
       "  cdd_name tigr tigrfam_name conj_pl_domains  \n",
       "0      NaN  NaN          NaN             NaN  \n",
       "1      NaN  NaN          NaN             NaN  \n",
       "2      NaN  NaN          NaN             NaN  \n",
       "3      NaN  NaN          NaN             NaN  \n",
       "4      NaN  NaN          NaN             NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "907343c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     3186\n",
       "unique     685\n",
       "top          0\n",
       "freq       354\n",
       "Name: cluster_num, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_data.cluster_num.describe()  # check-in"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "042ec120",
   "metadata": {},
   "source": [
    "**Read sequence data:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d80e8eeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>MAMSNMTYNNVFNHAYEMLKENIRYDDIRDTDDLHDAIHMTADNAV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>MSTTNVQYGLTAQTVFFYSDMVRCGFNWSLAMAQLKELYENNKAIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>MLTIGLLTVLDLAVGASFGKALGVAVGSYFTACIIIEIIKGALRK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>MMKHYVMPIHTSNGTTVCTPDGFAMKQRIERLKRELRINRKFFEGI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>MNTINIAKNDFSDIELAAIPFNTLADHYGERLAREQLALEHESYEM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID                                           sequence\n",
       "0  GCA_002600005.1_00001  MAMSNMTYNNVFNHAYEMLKENIRYDDIRDTDDLHDAIHMTADNAV...\n",
       "1  GCA_002600005.1_00002  MSTTNVQYGLTAQTVFFYSDMVRCGFNWSLAMAQLKELYENNKAIA...\n",
       "2  GCA_002600005.1_00003      MLTIGLLTVLDLAVGASFGKALGVAVGSYFTACIIIEIIKGALRK\n",
       "3  GCA_002600005.1_00004  MMKHYVMPIHTSNGTTVCTPDGFAMKQRIERLKRELRINRKFFEGI...\n",
       "4  GCA_002600005.1_00005  MNTINIAKNDFSDIELAAIPFNTLADHYGERLAREQLALEHESYEM..."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(faa_file, 'r') as fasta_file:\n",
    "    records = list(SimpleFastaParser(fasta_file))\n",
    "\n",
    "prot_seq_df = pd.DataFrame(records, columns = ['ID', 'sequence'])\n",
    "prot_seq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b8039af7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3186</td>\n",
       "      <td>3186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>3186</td>\n",
       "      <td>2803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>MSKLLATSKIEGQCTVTLREYYHGSMGSTYVVRYGKQVTHWVNPIL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           ID   \n",
       "count                    3186  \\\n",
       "unique                   3186   \n",
       "top     GCA_002600005.1_00001   \n",
       "freq                        1   \n",
       "\n",
       "                                                 sequence  \n",
       "count                                                3186  \n",
       "unique                                               2803  \n",
       "top     MSKLLATSKIEGQCTVTLREYYHGSMGSTYVVRYGKQVTHWVNPIL...  \n",
       "freq                                                   11  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_seq_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41a563bb",
   "metadata": {},
   "source": [
    "**Calculate sequence metrics: length, pI for each sequence.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d165087d",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_seq_df['pI'] = prot_seq_df['sequence'].apply(lambda x: IP(x).pi())\n",
    "prot_seq_df['pI'] = prot_seq_df['pI'].apply(lambda x: round(x, 2))\n",
    "prot_seq_df['length'] = prot_seq_df['sequence'].apply(lambda x: len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "07482112",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pI</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>MAMSNMTYNNVFNHAYEMLKENIRYDDIRDTDDLHDAIHMTADNAV...</td>\n",
       "      <td>4.05</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>MSTTNVQYGLTAQTVFFYSDMVRCGFNWSLAMAQLKELYENNKAIA...</td>\n",
       "      <td>4.75</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>MLTIGLLTVLDLAVGASFGKALGVAVGSYFTACIIIEIIKGALRK</td>\n",
       "      <td>9.10</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>MMKHYVMPIHTSNGTTVCTPDGFAMKQRIERLKRELRINRKFFEGI...</td>\n",
       "      <td>10.32</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>MNTINIAKNDFSDIELAAIPFNTLADHYGERLAREQLALEHESYEM...</td>\n",
       "      <td>7.59</td>\n",
       "      <td>883</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID                                           sequence   \n",
       "0  GCA_002600005.1_00001  MAMSNMTYNNVFNHAYEMLKENIRYDDIRDTDDLHDAIHMTADNAV...  \\\n",
       "1  GCA_002600005.1_00002  MSTTNVQYGLTAQTVFFYSDMVRCGFNWSLAMAQLKELYENNKAIA...   \n",
       "2  GCA_002600005.1_00003      MLTIGLLTVLDLAVGASFGKALGVAVGSYFTACIIIEIIKGALRK   \n",
       "3  GCA_002600005.1_00004  MMKHYVMPIHTSNGTTVCTPDGFAMKQRIERLKRELRINRKFFEGI...   \n",
       "4  GCA_002600005.1_00005  MNTINIAKNDFSDIELAAIPFNTLADHYGERLAREQLALEHESYEM...   \n",
       "\n",
       "      pI  length  \n",
       "0   4.05     117  \n",
       "1   4.75      51  \n",
       "2   9.10      45  \n",
       "3  10.32     214  \n",
       "4   7.59     883  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_seq_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7073d82d",
   "metadata": {},
   "source": [
    "**Aggregate results.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "065760ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>file_id</th>\n",
       "      <th>inference</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>...</th>\n",
       "      <th>tigr</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pI</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>is_Escherichia</th>\n",
       "      <th>is_Enterobacterales</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>901</td>\n",
       "      <td>1254</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>Overcome classical restriction gp0.3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>MAMSNMTYNNVFNHAYEMLKENIRYDDIRDTDDLHDAIHMTADNAV...</td>\n",
       "      <td>4.05</td>\n",
       "      <td>117</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1254</td>\n",
       "      <td>1409</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>Gene 0.4 protein</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>MSTTNVQYGLTAQTVFFYSDMVRCGFNWSLAMAQLKELYENNKAIA...</td>\n",
       "      <td>4.75</td>\n",
       "      <td>51</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1478</td>\n",
       "      <td>1615</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>MLTIGLLTVLDLAVGASFGKALGVAVGSYFTACIIIEIIKGALRK</td>\n",
       "      <td>9.10</td>\n",
       "      <td>45</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1612</td>\n",
       "      <td>2256</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>Protein kinase 0.7</td>\n",
       "      <td>134</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>MMKHYVMPIHTSNGTTVCTPDGFAMKQRIERLKRELRINRKFFEGI...</td>\n",
       "      <td>10.32</td>\n",
       "      <td>214</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>2327</td>\n",
       "      <td>4978</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>T7 RNA polymerase</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>MNTINIAKNDFSDIELAAIPFNTLADHYGERLAREQLALEHESYEM...</td>\n",
       "      <td>7.59</td>\n",
       "      <td>883</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id  start   end strand                     ID   \n",
       "0  AY264775.1    901  1254      +  GCA_002600005.1_00001  \\\n",
       "1  AY264775.1   1254  1409      +  GCA_002600005.1_00002   \n",
       "2  AY264775.1   1478  1615      +  GCA_002600005.1_00003   \n",
       "3  AY264775.1   1612  2256      +  GCA_002600005.1_00004   \n",
       "4  AY264775.1   2327  4978      +  GCA_002600005.1_00005   \n",
       "\n",
       "                   file_id                                          inference   \n",
       "0  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...  \\\n",
       "1  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "2  upstreams_with_clusters                  ab initio prediction:Prodigal:2.6   \n",
       "3  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "4  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "\n",
       "               locus_tag                               product cluster_num   \n",
       "0  GCA_002600005.1_00001  Overcome classical restriction gp0.3           2  \\\n",
       "1  GCA_002600005.1_00002                      Gene 0.4 protein          12   \n",
       "2  GCA_002600005.1_00003                  hypothetical protein          18   \n",
       "3  GCA_002600005.1_00004                    Protein kinase 0.7         134   \n",
       "4  GCA_002600005.1_00005                     T7 RNA polymerase           0   \n",
       "\n",
       "   ... tigr tigrfam_name conj_pl_domains                     ID   \n",
       "0  ...  NaN          NaN             NaN  GCA_002600005.1_00001  \\\n",
       "1  ...  NaN          NaN             NaN  GCA_002600005.1_00002   \n",
       "2  ...  NaN          NaN             NaN  GCA_002600005.1_00003   \n",
       "3  ...  NaN          NaN             NaN  GCA_002600005.1_00004   \n",
       "4  ...  NaN          NaN             NaN  GCA_002600005.1_00005   \n",
       "\n",
       "                                            sequence     pI length   \n",
       "0  MAMSNMTYNNVFNHAYEMLKENIRYDDIRDTDDLHDAIHMTADNAV...   4.05    117  \\\n",
       "1  MSTTNVQYGLTAQTVFFYSDMVRCGFNWSLAMAQLKELYENNKAIA...   4.75     51   \n",
       "2      MLTIGLLTVLDLAVGASFGKALGVAVGSYFTACIIIEIIKGALRK   9.10     45   \n",
       "3  MMKHYVMPIHTSNGTTVCTPDGFAMKQRIERLKRELRINRKFFEGI...  10.32    214   \n",
       "4  MNTINIAKNDFSDIELAAIPFNTLADHYGERLAREQLALEHESYEM...   7.59    883   \n",
       "\n",
       "              order is_Escherichia is_Enterobacterales  \n",
       "0  Enterobacterales          False                True  \n",
       "1  Enterobacterales          False                True  \n",
       "2  Enterobacterales          False                True  \n",
       "3  Enterobacterales          False                True  \n",
       "4  Enterobacterales          False                True  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([gff_data, prot_seq_df], axis=1)\n",
    "df = df.merge(hosts_data, on='seq_id') \n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b07d3334",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_agg = df.groupby('cluster_num').agg(\n",
    "                        MedianpI = pd.NamedAgg(column = 'pI', aggfunc=np.median), \n",
    "                        LQpI = pd.NamedAgg(column = 'pI', aggfunc = lambda x: x.quantile(0.25)),\n",
    "                        UQpI = pd.NamedAgg(column = 'pI', aggfunc = lambda x: x.quantile(0.75)),\n",
    "                        MedianLength = pd.NamedAgg(column = 'length', aggfunc=np.median), \n",
    "                        LQLenths = pd.NamedAgg(column = 'length', aggfunc = lambda x: x.quantile(0.25)),\n",
    "                        UQLenths = pd.NamedAgg(column = 'length', aggfunc = lambda x: x.quantile(0.75)),\n",
    "                        Ecoli_host = pd.NamedAgg(column = 'is_Escherichia', \n",
    "                                                 aggfunc = lambda x: x.any()),\n",
    "                        Enterobacter_host = pd.NamedAgg(column = 'is_Enterobacterales', \n",
    "                                                 aggfunc = lambda x: x.any())\n",
    "                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ad963af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_agg.index = df_agg.index.astype(int)\n",
    "df_agg.sort_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "47d1785e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_num = pd.Series(range(1, len(df_agg)+1))\n",
    "df_agg.set_index(cluster_num, inplace=True)\n",
    "df_agg['cluster_num'] = pd.Series(range(0, len(df_agg)+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "89142b29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MedianpI</th>\n",
       "      <th>LQpI</th>\n",
       "      <th>UQpI</th>\n",
       "      <th>MedianLength</th>\n",
       "      <th>LQLenths</th>\n",
       "      <th>UQLenths</th>\n",
       "      <th>Ecoli_host</th>\n",
       "      <th>Enterobacter_host</th>\n",
       "      <th>cluster_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6.960</td>\n",
       "      <td>6.7100</td>\n",
       "      <td>7.1500</td>\n",
       "      <td>888.0</td>\n",
       "      <td>883.0</td>\n",
       "      <td>906.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6.650</td>\n",
       "      <td>6.2675</td>\n",
       "      <td>7.5225</td>\n",
       "      <td>359.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>369.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.050</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>117.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10.780</td>\n",
       "      <td>10.5000</td>\n",
       "      <td>10.8200</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5.515</td>\n",
       "      <td>5.2200</td>\n",
       "      <td>6.8200</td>\n",
       "      <td>152.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   MedianpI     LQpI     UQpI  MedianLength  LQLenths  UQLenths  Ecoli_host   \n",
       "1     6.960   6.7100   7.1500         888.0     883.0     906.0        True  \\\n",
       "2     6.650   6.2675   7.5225         359.0     337.0     369.0        True   \n",
       "3     4.050   4.0500   4.0500         117.0     114.0     121.0        True   \n",
       "4    10.780  10.5000  10.8200          65.0      65.0      65.0        True   \n",
       "5     5.515   5.2200   6.8200         152.0     152.0     155.0        True   \n",
       "\n",
       "   Enterobacter_host  cluster_num  \n",
       "1               True            1  \n",
       "2               True            2  \n",
       "3               True            3  \n",
       "4               True            4  \n",
       "5               True            5  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_agg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb5bba21",
   "metadata": {},
   "source": [
    "**Merge results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "975f37b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_table = res_table.merge(df_agg, on='cluster_num', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ef141361",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clu_parent</th>\n",
       "      <th>count</th>\n",
       "      <th>products</th>\n",
       "      <th>sequence</th>\n",
       "      <th>phrOG</th>\n",
       "      <th>phrOG_annotation</th>\n",
       "      <th>phrOG_category</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "      <th>...</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>MedianpI</th>\n",
       "      <th>LQpI</th>\n",
       "      <th>UQpI</th>\n",
       "      <th>MedianLength</th>\n",
       "      <th>LQLenths</th>\n",
       "      <th>UQLenths</th>\n",
       "      <th>Ecoli_host</th>\n",
       "      <th>Enterobacter_host</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_001470995.1_00023</td>\n",
       "      <td>354</td>\n",
       "      <td>{T7 RNA polymerase: 354}</td>\n",
       "      <td>MEYNPLTELASLYGEDLAAEQLRLEAEAYSLGEKRFMEAMEFKAET...</td>\n",
       "      <td>phrog_20599,phrog_414</td>\n",
       "      <td>RNA polymerase</td>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>6.960</td>\n",
       "      <td>6.7100</td>\n",
       "      <td>7.1500</td>\n",
       "      <td>888.0</td>\n",
       "      <td>883.0</td>\n",
       "      <td>906.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_025085155.1_00014</td>\n",
       "      <td>190</td>\n",
       "      <td>{Protein kinase 0.7: 183, hypothetical protein...</td>\n",
       "      <td>MYAHFQSLLKAIRELPINELDKRQPMLVDLLAQIVNHETQDGAITA...</td>\n",
       "      <td>phrog_18305,phrog_2828,phrog_15752</td>\n",
       "      <td>kinase,serine-threonine kinase,protein kinase;...</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>6.650</td>\n",
       "      <td>6.2675</td>\n",
       "      <td>7.5225</td>\n",
       "      <td>359.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>369.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_001041815.1_00006</td>\n",
       "      <td>150</td>\n",
       "      <td>{Overcome classical restriction gp0.3: 102, hy...</td>\n",
       "      <td>MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...</td>\n",
       "      <td>phrog_2568</td>\n",
       "      <td>ocr-like anti-restriction</td>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>4.050</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>117.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_023716895.1_00035</td>\n",
       "      <td>149</td>\n",
       "      <td>{hypothetical protein: 149}</td>\n",
       "      <td>METSCRLPHGAGRSCTGMDGDAMMFTAFIVFMYLLIVLYFLKDFRK...</td>\n",
       "      <td>phrog_2385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>10.780</td>\n",
       "      <td>10.5000</td>\n",
       "      <td>10.8200</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_009672325.1_00007</td>\n",
       "      <td>144</td>\n",
       "      <td>{hypothetical protein: 144}</td>\n",
       "      <td>MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>5.515</td>\n",
       "      <td>5.2200</td>\n",
       "      <td>6.8200</td>\n",
       "      <td>152.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              clu_parent  count   \n",
       "0  GCF_001470995.1_00023    354  \\\n",
       "1  GCA_025085155.1_00014    190   \n",
       "2  GCF_001041815.1_00006    150   \n",
       "3  GCA_023716895.1_00035    149   \n",
       "4  GCA_009672325.1_00007    144   \n",
       "\n",
       "                                            products   \n",
       "0                           {T7 RNA polymerase: 354}  \\\n",
       "1  {Protein kinase 0.7: 183, hypothetical protein...   \n",
       "2  {Overcome classical restriction gp0.3: 102, hy...   \n",
       "3                        {hypothetical protein: 149}   \n",
       "4                        {hypothetical protein: 144}   \n",
       "\n",
       "                                            sequence   \n",
       "0  MEYNPLTELASLYGEDLAAEQLRLEAEAYSLGEKRFMEAMEFKAET...  \\\n",
       "1  MYAHFQSLLKAIRELPINELDKRQPMLVDLLAQIVNHETQDGAITA...   \n",
       "2  MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...   \n",
       "3  METSCRLPHGAGRSCTGMDGDAMMFTAFIVFMYLLIVLYFLKDFRK...   \n",
       "4  MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...   \n",
       "\n",
       "                                phrOG   \n",
       "0               phrog_20599,phrog_414  \\\n",
       "1  phrog_18305,phrog_2828,phrog_15752   \n",
       "2                          phrog_2568   \n",
       "3                          phrog_2385   \n",
       "4                          phrog_2226   \n",
       "\n",
       "                                    phrOG_annotation   \n",
       "0                                     RNA polymerase  \\\n",
       "1  kinase,serine-threonine kinase,protein kinase;...   \n",
       "2                          ocr-like anti-restriction   \n",
       "3                                                NaN   \n",
       "4                    SAM-dependent methyltransferase   \n",
       "\n",
       "                                      phrOG_category pfam_annotation   \n",
       "0                 DNA, RNA and nucleotide metabolism  RNA_pol,RPOL_N  \\\n",
       "1                                              other             NaN   \n",
       "2  moron, auxiliary metabolic gene and host takeover             ocr   \n",
       "3                                   unknown function             NaN   \n",
       "4                                              other             NaN   \n",
       "\n",
       "                   pfam conj_pl_domains  ... tigrfam_name cluster_num   \n",
       "0  PF00940.22,PF14700.9             NaN  ...          NaN           1  \\\n",
       "1                   NaN             NaN  ...          NaN           2   \n",
       "2            PF08684.13             NaN  ...          NaN           3   \n",
       "3                   NaN             NaN  ...          NaN           4   \n",
       "4                   NaN             NaN  ...          NaN           5   \n",
       "\n",
       "  MedianpI     LQpI     UQpI  MedianLength  LQLenths  UQLenths  Ecoli_host   \n",
       "0    6.960   6.7100   7.1500         888.0     883.0     906.0        True  \\\n",
       "1    6.650   6.2675   7.5225         359.0     337.0     369.0        True   \n",
       "2    4.050   4.0500   4.0500         117.0     114.0     121.0        True   \n",
       "3   10.780  10.5000  10.8200          65.0      65.0      65.0        True   \n",
       "4    5.515   5.2200   6.8200         152.0     152.0     155.0        True   \n",
       "\n",
       "   Enterobacter_host  \n",
       "0               True  \n",
       "1               True  \n",
       "2               True  \n",
       "3               True  \n",
       "4               True  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "57f236d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "last_seq = res_table.iloc[-1, 3]\n",
    "last_pi = round(IP(last_seq).pi(), 2)\n",
    "last_length = len(last_seq)\n",
    "res_table.iloc[-1, 18] = last_length\n",
    "res_table.iloc[-1, 15] = last_pi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e22e6e5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# res_table.to_csv(output_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4a6eb98f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clu_parent</th>\n",
       "      <th>count</th>\n",
       "      <th>products</th>\n",
       "      <th>sequence</th>\n",
       "      <th>phrOG</th>\n",
       "      <th>phrOG_annotation</th>\n",
       "      <th>phrOG_category</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "      <th>...</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>MedianpI</th>\n",
       "      <th>LQpI</th>\n",
       "      <th>UQpI</th>\n",
       "      <th>MedianLength</th>\n",
       "      <th>LQLenths</th>\n",
       "      <th>UQLenths</th>\n",
       "      <th>Ecoli_host</th>\n",
       "      <th>Enterobacter_host</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_001041815.1_00006</td>\n",
       "      <td>150</td>\n",
       "      <td>{Overcome classical restriction gp0.3: 102, hy...</td>\n",
       "      <td>MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...</td>\n",
       "      <td>phrog_2568</td>\n",
       "      <td>ocr-like anti-restriction</td>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>4.050</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>117.0</td>\n",
       "      <td>114.00</td>\n",
       "      <td>121.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_009672325.1_00007</td>\n",
       "      <td>144</td>\n",
       "      <td>{hypothetical protein: 144}</td>\n",
       "      <td>MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>5.515</td>\n",
       "      <td>5.2200</td>\n",
       "      <td>6.8200</td>\n",
       "      <td>152.0</td>\n",
       "      <td>152.00</td>\n",
       "      <td>155.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCA_021863795.1_00008</td>\n",
       "      <td>53</td>\n",
       "      <td>{hypothetical protein: 53}</td>\n",
       "      <td>MTTNTSVITQEAGETVSIMSFREEVALEIRSHLDNIGQSYLAVGKA...</td>\n",
       "      <td>phrog_4571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>4.920</td>\n",
       "      <td>4.8500</td>\n",
       "      <td>5.0200</td>\n",
       "      <td>297.0</td>\n",
       "      <td>294.00</td>\n",
       "      <td>304.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>GCA_009667685.1_00049</td>\n",
       "      <td>17</td>\n",
       "      <td>{hypothetical protein: 17}</td>\n",
       "      <td>MKLIGKGAFTKAYLLDSGRVQLHTCCPIKESMAWGWFPDSDLFPKL...</td>\n",
       "      <td>phrog_11084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31</td>\n",
       "      <td>5.630</td>\n",
       "      <td>5.4100</td>\n",
       "      <td>5.9400</td>\n",
       "      <td>174.0</td>\n",
       "      <td>172.00</td>\n",
       "      <td>174.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>GCF_000917035.1_00005</td>\n",
       "      <td>12</td>\n",
       "      <td>{Protein kinase 0.7: 11, hypothetical protein: 1}</td>\n",
       "      <td>MKDTMHRLTALLHEVMVAADYRVNCSGMGCAGQDSWVSFEAMARGE...</td>\n",
       "      <td>phrog_18305,phrog_2828,phrog_15752</td>\n",
       "      <td>kinase,serine-threonine kinase,protein kinase;...</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>5.475</td>\n",
       "      <td>5.3500</td>\n",
       "      <td>5.6800</td>\n",
       "      <td>244.0</td>\n",
       "      <td>244.00</td>\n",
       "      <td>244.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GCA_021863865.1_00009</td>\n",
       "      <td>8</td>\n",
       "      <td>{hypothetical protein: 8}</td>\n",
       "      <td>MQKLIDALNASVFTLRTKGKATIKAEVAMLTEAGVTCDELDTSSKA...</td>\n",
       "      <td>phrog_4571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59</td>\n",
       "      <td>5.000</td>\n",
       "      <td>4.9800</td>\n",
       "      <td>5.0425</td>\n",
       "      <td>344.0</td>\n",
       "      <td>338.75</td>\n",
       "      <td>344.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>GCF_002743475.1_00004</td>\n",
       "      <td>4</td>\n",
       "      <td>{hypothetical protein: 4}</td>\n",
       "      <td>MLNRIDVHFKLLSAEKDIYFKRTWTTEPSDYPFIMGINIDGTMYYP...</td>\n",
       "      <td>phrog_23478</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>91</td>\n",
       "      <td>4.360</td>\n",
       "      <td>4.3200</td>\n",
       "      <td>4.4025</td>\n",
       "      <td>166.0</td>\n",
       "      <td>166.00</td>\n",
       "      <td>166.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>GCA_021863795.1_00004</td>\n",
       "      <td>4</td>\n",
       "      <td>{hypothetical protein: 4}</td>\n",
       "      <td>MIAQTIKKAKYTVIASASRSNLTEAANIDRHIDAGFTLSALGYDRH...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>104</td>\n",
       "      <td>5.650</td>\n",
       "      <td>5.5200</td>\n",
       "      <td>5.7800</td>\n",
       "      <td>143.0</td>\n",
       "      <td>143.00</td>\n",
       "      <td>143.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>GCA_020488535.1_00041</td>\n",
       "      <td>4</td>\n",
       "      <td>{hypothetical protein: 4}</td>\n",
       "      <td>MERNANAYYDLLSATVELFRERIEQDQLTEDNYHDALHEVVDGQVP...</td>\n",
       "      <td>phrog_2568</td>\n",
       "      <td>ocr-like anti-restriction</td>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107</td>\n",
       "      <td>4.050</td>\n",
       "      <td>4.0500</td>\n",
       "      <td>4.0525</td>\n",
       "      <td>172.5</td>\n",
       "      <td>172.00</td>\n",
       "      <td>175.50</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>GCF_002958225.1_00005</td>\n",
       "      <td>3</td>\n",
       "      <td>{hypothetical protein: 3}</td>\n",
       "      <td>MRNLERAMQLATLTLDNFSNWFEDEGGVSHFVMSKVKMDEGTSQCE...</td>\n",
       "      <td>phrog_18383,phrog_23974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115</td>\n",
       "      <td>4.620</td>\n",
       "      <td>4.5950</td>\n",
       "      <td>4.6550</td>\n",
       "      <td>108.0</td>\n",
       "      <td>108.00</td>\n",
       "      <td>108.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>GCF_001745895.1_00008</td>\n",
       "      <td>3</td>\n",
       "      <td>{hypothetical protein: 3}</td>\n",
       "      <td>MINKAPTTIINAGSYLEGLRDKFKLDVKLIAQKKHPKVYLLSHEEI...</td>\n",
       "      <td>phrog_31405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>117</td>\n",
       "      <td>5.790</td>\n",
       "      <td>5.5000</td>\n",
       "      <td>6.3050</td>\n",
       "      <td>170.0</td>\n",
       "      <td>169.00</td>\n",
       "      <td>170.50</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>GCA_021870105.1_00001</td>\n",
       "      <td>3</td>\n",
       "      <td>{hypothetical protein: 2, Protein kinase 0.7: 1}</td>\n",
       "      <td>MAMSNMTYNNVFDHAYEMLKENIRYDDIRDTDDLHDAIHMAADNAV...</td>\n",
       "      <td>phrog_2828,phrog_2568</td>\n",
       "      <td>serine-threonine kinase,ocr-like anti-restriction</td>\n",
       "      <td>other,moron, auxiliary metabolic gene and host...</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>135</td>\n",
       "      <td>5.570</td>\n",
       "      <td>5.5700</td>\n",
       "      <td>7.9450</td>\n",
       "      <td>231.0</td>\n",
       "      <td>222.50</td>\n",
       "      <td>231.00</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>GCF_001041455.1_00006</td>\n",
       "      <td>2</td>\n",
       "      <td>{hypothetical protein: 2}</td>\n",
       "      <td>MALVENFTSRNGNKVANQFKLFLAHGTVFQSYTSLIATKLYDGRVY...</td>\n",
       "      <td>phrog_21575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>189</td>\n",
       "      <td>5.560</td>\n",
       "      <td>5.3100</td>\n",
       "      <td>5.8100</td>\n",
       "      <td>108.0</td>\n",
       "      <td>104.50</td>\n",
       "      <td>111.50</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>GCF_000900955.1_00006</td>\n",
       "      <td>2</td>\n",
       "      <td>{hypothetical protein: 2}</td>\n",
       "      <td>MFKTNAKFIVLASAFRFGVSQDTNIENHVDVRFSLDAQGFQHMDVL...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>196</td>\n",
       "      <td>5.470</td>\n",
       "      <td>5.4700</td>\n",
       "      <td>5.4700</td>\n",
       "      <td>138.0</td>\n",
       "      <td>138.00</td>\n",
       "      <td>138.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>GCA_020488535.1_00046</td>\n",
       "      <td>2</td>\n",
       "      <td>{Terminase%2C large subunit gp19: 2}</td>\n",
       "      <td>VTTEQKQLERIKNNFVLFLFVLWKALNLPQPTRCQQDMAKTLANGQ...</td>\n",
       "      <td>phrog_17672,phrog_15524,phrog_38,phrog_4758</td>\n",
       "      <td>terminase large subunit</td>\n",
       "      <td>head and packaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>225</td>\n",
       "      <td>5.445</td>\n",
       "      <td>5.4275</td>\n",
       "      <td>5.4625</td>\n",
       "      <td>584.5</td>\n",
       "      <td>583.25</td>\n",
       "      <td>585.75</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>GCA_012637005.1_00017</td>\n",
       "      <td>2</td>\n",
       "      <td>{hypothetical protein: 2}</td>\n",
       "      <td>MLHTNEPANKIMVALSAYRATNSTQRNIESQVVLLNALKAGSFEIL...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>238</td>\n",
       "      <td>4.700</td>\n",
       "      <td>4.6650</td>\n",
       "      <td>4.7350</td>\n",
       "      <td>150.0</td>\n",
       "      <td>149.50</td>\n",
       "      <td>150.50</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>GCF_003094295.1_00001</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MVYLSAFRNERSEYENNKKHSNLVSWLQSQAFDRGNDFKVVSASLT...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>289</td>\n",
       "      <td>5.060</td>\n",
       "      <td>5.0600</td>\n",
       "      <td>5.0600</td>\n",
       "      <td>227.0</td>\n",
       "      <td>227.00</td>\n",
       "      <td>227.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>GCA_025630455.1_00057</td>\n",
       "      <td>1</td>\n",
       "      <td>{Tail spike protein: 1}</td>\n",
       "      <td>MAPEDDRPGDLPVSQYPMHQLPNNHMVDNILVMNSLGVGLGMDGRD...</td>\n",
       "      <td>phrog_1819,phrog_23354</td>\n",
       "      <td>tail spike protein</td>\n",
       "      <td>tail</td>\n",
       "      <td>PhageP22-tail</td>\n",
       "      <td>PF09251.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>478</td>\n",
       "      <td>5.420</td>\n",
       "      <td>5.4200</td>\n",
       "      <td>5.4200</td>\n",
       "      <td>262.0</td>\n",
       "      <td>262.00</td>\n",
       "      <td>262.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>GCA_015502225.1_00041</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MALTITTKYLSVTNTRGARIKVSCSNGVSKTYSYDHQASEPHAVAI...</td>\n",
       "      <td>phrog_3199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>596</td>\n",
       "      <td>5.810</td>\n",
       "      <td>5.8100</td>\n",
       "      <td>5.8100</td>\n",
       "      <td>113.0</td>\n",
       "      <td>113.00</td>\n",
       "      <td>113.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>GCA_009828395.2_00002</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MMINRDTAFKALFEALKAKASLNTDGFLAAAMDDIEDTMGTIQPAN...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>625</td>\n",
       "      <td>4.370</td>\n",
       "      <td>4.3700</td>\n",
       "      <td>4.3700</td>\n",
       "      <td>142.0</td>\n",
       "      <td>142.00</td>\n",
       "      <td>142.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>GCA_009828395.2_00001</td>\n",
       "      <td>1</td>\n",
       "      <td>{hypothetical protein: 1}</td>\n",
       "      <td>MIHTKEPTNCFYVYLSAYRSMNDLSTNLKAHAKMAEWLGYQQYARD...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>626</td>\n",
       "      <td>5.830</td>\n",
       "      <td>5.8300</td>\n",
       "      <td>5.8300</td>\n",
       "      <td>159.0</td>\n",
       "      <td>159.00</td>\n",
       "      <td>159.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                clu_parent  count   \n",
       "2    GCF_001041815.1_00006    150  \\\n",
       "4    GCA_009672325.1_00007    144   \n",
       "9    GCA_021863795.1_00008     53   \n",
       "30   GCA_009667685.1_00049     17   \n",
       "39   GCF_000917035.1_00005     12   \n",
       "58   GCA_021863865.1_00009      8   \n",
       "90   GCF_002743475.1_00004      4   \n",
       "103  GCA_021863795.1_00004      4   \n",
       "106  GCA_020488535.1_00041      4   \n",
       "114  GCF_002958225.1_00005      3   \n",
       "116  GCF_001745895.1_00008      3   \n",
       "134  GCA_021870105.1_00001      3   \n",
       "188  GCF_001041455.1_00006      2   \n",
       "195  GCF_000900955.1_00006      2   \n",
       "224  GCA_020488535.1_00046      2   \n",
       "237  GCA_012637005.1_00017      2   \n",
       "288  GCF_003094295.1_00001      1   \n",
       "477  GCA_025630455.1_00057      1   \n",
       "595  GCA_015502225.1_00041      1   \n",
       "624  GCA_009828395.2_00002      1   \n",
       "625  GCA_009828395.2_00001      1   \n",
       "\n",
       "                                              products   \n",
       "2    {Overcome classical restriction gp0.3: 102, hy...  \\\n",
       "4                          {hypothetical protein: 144}   \n",
       "9                           {hypothetical protein: 53}   \n",
       "30                          {hypothetical protein: 17}   \n",
       "39   {Protein kinase 0.7: 11, hypothetical protein: 1}   \n",
       "58                           {hypothetical protein: 8}   \n",
       "90                           {hypothetical protein: 4}   \n",
       "103                          {hypothetical protein: 4}   \n",
       "106                          {hypothetical protein: 4}   \n",
       "114                          {hypothetical protein: 3}   \n",
       "116                          {hypothetical protein: 3}   \n",
       "134   {hypothetical protein: 2, Protein kinase 0.7: 1}   \n",
       "188                          {hypothetical protein: 2}   \n",
       "195                          {hypothetical protein: 2}   \n",
       "224               {Terminase%2C large subunit gp19: 2}   \n",
       "237                          {hypothetical protein: 2}   \n",
       "288                          {hypothetical protein: 1}   \n",
       "477                            {Tail spike protein: 1}   \n",
       "595                          {hypothetical protein: 1}   \n",
       "624                          {hypothetical protein: 1}   \n",
       "625                          {hypothetical protein: 1}   \n",
       "\n",
       "                                              sequence   \n",
       "2    MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...  \\\n",
       "4    MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...   \n",
       "9    MTTNTSVITQEAGETVSIMSFREEVALEIRSHLDNIGQSYLAVGKA...   \n",
       "30   MKLIGKGAFTKAYLLDSGRVQLHTCCPIKESMAWGWFPDSDLFPKL...   \n",
       "39   MKDTMHRLTALLHEVMVAADYRVNCSGMGCAGQDSWVSFEAMARGE...   \n",
       "58   MQKLIDALNASVFTLRTKGKATIKAEVAMLTEAGVTCDELDTSSKA...   \n",
       "90   MLNRIDVHFKLLSAEKDIYFKRTWTTEPSDYPFIMGINIDGTMYYP...   \n",
       "103  MIAQTIKKAKYTVIASASRSNLTEAANIDRHIDAGFTLSALGYDRH...   \n",
       "106  MERNANAYYDLLSATVELFRERIEQDQLTEDNYHDALHEVVDGQVP...   \n",
       "114  MRNLERAMQLATLTLDNFSNWFEDEGGVSHFVMSKVKMDEGTSQCE...   \n",
       "116  MINKAPTTIINAGSYLEGLRDKFKLDVKLIAQKKHPKVYLLSHEEI...   \n",
       "134  MAMSNMTYNNVFDHAYEMLKENIRYDDIRDTDDLHDAIHMAADNAV...   \n",
       "188  MALVENFTSRNGNKVANQFKLFLAHGTVFQSYTSLIATKLYDGRVY...   \n",
       "195  MFKTNAKFIVLASAFRFGVSQDTNIENHVDVRFSLDAQGFQHMDVL...   \n",
       "224  VTTEQKQLERIKNNFVLFLFVLWKALNLPQPTRCQQDMAKTLANGQ...   \n",
       "237  MLHTNEPANKIMVALSAYRATNSTQRNIESQVVLLNALKAGSFEIL...   \n",
       "288  MVYLSAFRNERSEYENNKKHSNLVSWLQSQAFDRGNDFKVVSASLT...   \n",
       "477  MAPEDDRPGDLPVSQYPMHQLPNNHMVDNILVMNSLGVGLGMDGRD...   \n",
       "595  MALTITTKYLSVTNTRGARIKVSCSNGVSKTYSYDHQASEPHAVAI...   \n",
       "624  MMINRDTAFKALFEALKAKASLNTDGFLAAAMDDIEDTMGTIQPAN...   \n",
       "625  MIHTKEPTNCFYVYLSAYRSMNDLSTNLKAHAKMAEWLGYQQYARD...   \n",
       "\n",
       "                                           phrOG   \n",
       "2                                     phrog_2568  \\\n",
       "4                                     phrog_2226   \n",
       "9                                     phrog_4571   \n",
       "30                                   phrog_11084   \n",
       "39            phrog_18305,phrog_2828,phrog_15752   \n",
       "58                                    phrog_4571   \n",
       "90                                   phrog_23478   \n",
       "103                                   phrog_2226   \n",
       "106                                   phrog_2568   \n",
       "114                      phrog_18383,phrog_23974   \n",
       "116                                  phrog_31405   \n",
       "134                        phrog_2828,phrog_2568   \n",
       "188                                  phrog_21575   \n",
       "195                                   phrog_2226   \n",
       "224  phrog_17672,phrog_15524,phrog_38,phrog_4758   \n",
       "237                                   phrog_2226   \n",
       "288                                   phrog_2226   \n",
       "477                       phrog_1819,phrog_23354   \n",
       "595                                   phrog_3199   \n",
       "624                                          NaN   \n",
       "625                                   phrog_2226   \n",
       "\n",
       "                                      phrOG_annotation   \n",
       "2                            ocr-like anti-restriction  \\\n",
       "4                      SAM-dependent methyltransferase   \n",
       "9                                                  NaN   \n",
       "30                                                 NaN   \n",
       "39   kinase,serine-threonine kinase,protein kinase;...   \n",
       "58                                                 NaN   \n",
       "90                                                 NaN   \n",
       "103                    SAM-dependent methyltransferase   \n",
       "106                          ocr-like anti-restriction   \n",
       "114                                                NaN   \n",
       "116                                                NaN   \n",
       "134  serine-threonine kinase,ocr-like anti-restriction   \n",
       "188                                                NaN   \n",
       "195                    SAM-dependent methyltransferase   \n",
       "224                            terminase large subunit   \n",
       "237                    SAM-dependent methyltransferase   \n",
       "288                    SAM-dependent methyltransferase   \n",
       "477                                 tail spike protein   \n",
       "595                                                NaN   \n",
       "624                                                NaN   \n",
       "625                    SAM-dependent methyltransferase   \n",
       "\n",
       "                                        phrOG_category pfam_annotation   \n",
       "2    moron, auxiliary metabolic gene and host takeover             ocr  \\\n",
       "4                                                other             NaN   \n",
       "9                                     unknown function             NaN   \n",
       "30                                    unknown function             NaN   \n",
       "39                                               other             NaN   \n",
       "58                                    unknown function             NaN   \n",
       "90                                    unknown function             NaN   \n",
       "103                                              other             NaN   \n",
       "106  moron, auxiliary metabolic gene and host takeover             ocr   \n",
       "114                                   unknown function             NaN   \n",
       "116                                   unknown function             NaN   \n",
       "134  other,moron, auxiliary metabolic gene and host...             ocr   \n",
       "188                                   unknown function             NaN   \n",
       "195                                              other             NaN   \n",
       "224                                 head and packaging             NaN   \n",
       "237                                              other             NaN   \n",
       "288                                              other             NaN   \n",
       "477                                               tail   PhageP22-tail   \n",
       "595                                   unknown function             NaN   \n",
       "624                                                NaN             NaN   \n",
       "625                                              other             NaN   \n",
       "\n",
       "           pfam conj_pl_domains  ... tigrfam_name cluster_num MedianpI   \n",
       "2    PF08684.13             NaN  ...          NaN           3    4.050  \\\n",
       "4           NaN             NaN  ...          NaN           5    5.515   \n",
       "9           NaN             NaN  ...          NaN          10    4.920   \n",
       "30          NaN             NaN  ...          NaN          31    5.630   \n",
       "39          NaN             NaN  ...          NaN          40    5.475   \n",
       "58          NaN             NaN  ...          NaN          59    5.000   \n",
       "90          NaN             NaN  ...          NaN          91    4.360   \n",
       "103         NaN             NaN  ...          NaN         104    5.650   \n",
       "106  PF08684.13             NaN  ...          NaN         107    4.050   \n",
       "114         NaN             NaN  ...          NaN         115    4.620   \n",
       "116         NaN             NaN  ...          NaN         117    5.790   \n",
       "134  PF08684.13             NaN  ...          NaN         135    5.570   \n",
       "188         NaN             NaN  ...          NaN         189    5.560   \n",
       "195         NaN             NaN  ...          NaN         196    5.470   \n",
       "224         NaN             NaN  ...          NaN         225    5.445   \n",
       "237         NaN             NaN  ...          NaN         238    4.700   \n",
       "288         NaN             NaN  ...          NaN         289    5.060   \n",
       "477  PF09251.13             NaN  ...          NaN         478    5.420   \n",
       "595         NaN             NaN  ...          NaN         596    5.810   \n",
       "624         NaN             NaN  ...          NaN         625    4.370   \n",
       "625         NaN             NaN  ...          NaN         626    5.830   \n",
       "\n",
       "       LQpI    UQpI  MedianLength  LQLenths  UQLenths  Ecoli_host   \n",
       "2    4.0500  4.0500         117.0    114.00    121.00        True  \\\n",
       "4    5.2200  6.8200         152.0    152.00    155.00        True   \n",
       "9    4.8500  5.0200         297.0    294.00    304.00        True   \n",
       "30   5.4100  5.9400         174.0    172.00    174.00        True   \n",
       "39   5.3500  5.6800         244.0    244.00    244.00        True   \n",
       "58   4.9800  5.0425         344.0    338.75    344.00        True   \n",
       "90   4.3200  4.4025         166.0    166.00    166.00       False   \n",
       "103  5.5200  5.7800         143.0    143.00    143.00       False   \n",
       "106  4.0500  4.0525         172.5    172.00    175.50       False   \n",
       "114  4.5950  4.6550         108.0    108.00    108.00       False   \n",
       "116  5.5000  6.3050         170.0    169.00    170.50       False   \n",
       "134  5.5700  7.9450         231.0    222.50    231.00        True   \n",
       "188  5.3100  5.8100         108.0    104.50    111.50       False   \n",
       "195  5.4700  5.4700         138.0    138.00    138.00       False   \n",
       "224  5.4275  5.4625         584.5    583.25    585.75        True   \n",
       "237  4.6650  4.7350         150.0    149.50    150.50       False   \n",
       "288  5.0600  5.0600         227.0    227.00    227.00       False   \n",
       "477  5.4200  5.4200         262.0    262.00    262.00       False   \n",
       "595  5.8100  5.8100         113.0    113.00    113.00       False   \n",
       "624  4.3700  4.3700         142.0    142.00    142.00       False   \n",
       "625  5.8300  5.8300         159.0    159.00    159.00       False   \n",
       "\n",
       "     Enterobacter_host  \n",
       "2                 True  \n",
       "4                 True  \n",
       "9                 True  \n",
       "30                True  \n",
       "39                True  \n",
       "58                True  \n",
       "90                True  \n",
       "103               True  \n",
       "106               True  \n",
       "114               True  \n",
       "116               True  \n",
       "134               True  \n",
       "188               True  \n",
       "195               True  \n",
       "224              False  \n",
       "237               True  \n",
       "288               True  \n",
       "477               True  \n",
       "595               True  \n",
       "624               True  \n",
       "625               True  \n",
       "\n",
       "[21 rows x 23 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_table.query('MedianpI < 6 and MedianLength > 100 and (Ecoli_host or Enterobacter_host)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1ffd1a5",
   "metadata": {},
   "source": [
    "Next goal is to find, **which gene lay near the Ocr or SAMase or at the place.**\n",
    "\n",
    "We need to find neighbours."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "219e9b6e",
   "metadata": {},
   "source": [
    "**Find SAMase and Ocr neighbours**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e44b526f",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_ocr = set([3, 107, 135, 297])\n",
    "clusters_samase = set([5, 33, 55, 104, 196, 238, 289, 308, 372, 503, 626])\n",
    "clusters_ard = set([256, 382, 488, 583])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5b45389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>file_id</th>\n",
       "      <th>inference</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>...</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>e_c_number</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdd</th>\n",
       "      <th>cdd_name</th>\n",
       "      <th>tigr</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>901</td>\n",
       "      <td>1254</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>Overcome classical restriction gp0.3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1254</td>\n",
       "      <td>1409</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>Gene 0.4 protein</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1478</td>\n",
       "      <td>1615</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1612</td>\n",
       "      <td>2256</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>Protein kinase 0.7</td>\n",
       "      <td>134</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.11.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>2327</td>\n",
       "      <td>4978</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>T7 RNA polymerase</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>2.7.7.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id  start   end strand                     ID   \n",
       "0  AY264775.1    901  1254      +  GCA_002600005.1_00001  \\\n",
       "1  AY264775.1   1254  1409      +  GCA_002600005.1_00002   \n",
       "2  AY264775.1   1478  1615      +  GCA_002600005.1_00003   \n",
       "3  AY264775.1   1612  2256      +  GCA_002600005.1_00004   \n",
       "4  AY264775.1   2327  4978      +  GCA_002600005.1_00005   \n",
       "\n",
       "                   file_id                                          inference   \n",
       "0  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...  \\\n",
       "1  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "2  upstreams_with_clusters                  ab initio prediction:Prodigal:2.6   \n",
       "3  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "4  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "\n",
       "               locus_tag                               product cluster_num   \n",
       "0  GCA_002600005.1_00001  Overcome classical restriction gp0.3           2  \\\n",
       "1  GCA_002600005.1_00002                      Gene 0.4 protein          12   \n",
       "2  GCA_002600005.1_00003                  hypothetical protein          18   \n",
       "3  GCA_002600005.1_00004                    Protein kinase 0.7         134   \n",
       "4  GCA_002600005.1_00005                     T7 RNA polymerase           0   \n",
       "\n",
       "   ... pfam_annotation                  pfam e_c_number Name gene  cdd   \n",
       "0  ...             ocr            PF08684.13        NaN  NaN  NaN  NaN  \\\n",
       "1  ...             NaN                   NaN        NaN  NaN  NaN  NaN   \n",
       "2  ...             NaN                   NaN        NaN  NaN  NaN  NaN   \n",
       "3  ...             NaN                   NaN   2.7.11.1  NaN  NaN  NaN   \n",
       "4  ...  RNA_pol,RPOL_N  PF00940.22,PF14700.9    2.7.7.6    1    1  NaN   \n",
       "\n",
       "  cdd_name tigr tigrfam_name conj_pl_domains  \n",
       "0      NaN  NaN          NaN             NaN  \n",
       "1      NaN  NaN          NaN             NaN  \n",
       "2      NaN  NaN          NaN             NaN  \n",
       "3      NaN  NaN          NaN             NaN  \n",
       "4      NaN  NaN          NaN             NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5f4d7f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_data.cluster_num = gff_data.cluster_num.apply(lambda x: int(x) + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0ee1f45b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ID</th>\n",
       "      <th>file_id</th>\n",
       "      <th>inference</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>cluster_num</th>\n",
       "      <th>...</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>e_c_number</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdd</th>\n",
       "      <th>cdd_name</th>\n",
       "      <th>tigr</th>\n",
       "      <th>tigrfam_name</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>901</td>\n",
       "      <td>1254</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00001</td>\n",
       "      <td>Overcome classical restriction gp0.3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1254</td>\n",
       "      <td>1409</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00002</td>\n",
       "      <td>Gene 0.4 protein</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1478</td>\n",
       "      <td>1615</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6</td>\n",
       "      <td>GCA_002600005.1_00003</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>1612</td>\n",
       "      <td>2256</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00004</td>\n",
       "      <td>Protein kinase 0.7</td>\n",
       "      <td>135</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.7.11.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AY264775.1</td>\n",
       "      <td>2327</td>\n",
       "      <td>4978</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>upstreams_with_clusters</td>\n",
       "      <td>ab initio prediction:Prodigal:2.6,similar to A...</td>\n",
       "      <td>GCA_002600005.1_00005</td>\n",
       "      <td>T7 RNA polymerase</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>2.7.7.6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id  start   end strand                     ID   \n",
       "0  AY264775.1    901  1254      +  GCA_002600005.1_00001  \\\n",
       "1  AY264775.1   1254  1409      +  GCA_002600005.1_00002   \n",
       "2  AY264775.1   1478  1615      +  GCA_002600005.1_00003   \n",
       "3  AY264775.1   1612  2256      +  GCA_002600005.1_00004   \n",
       "4  AY264775.1   2327  4978      +  GCA_002600005.1_00005   \n",
       "\n",
       "                   file_id                                          inference   \n",
       "0  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...  \\\n",
       "1  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "2  upstreams_with_clusters                  ab initio prediction:Prodigal:2.6   \n",
       "3  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "4  upstreams_with_clusters  ab initio prediction:Prodigal:2.6,similar to A...   \n",
       "\n",
       "               locus_tag                               product  cluster_num   \n",
       "0  GCA_002600005.1_00001  Overcome classical restriction gp0.3            3  \\\n",
       "1  GCA_002600005.1_00002                      Gene 0.4 protein           13   \n",
       "2  GCA_002600005.1_00003                  hypothetical protein           19   \n",
       "3  GCA_002600005.1_00004                    Protein kinase 0.7          135   \n",
       "4  GCA_002600005.1_00005                     T7 RNA polymerase            1   \n",
       "\n",
       "   ... pfam_annotation                  pfam e_c_number Name gene  cdd   \n",
       "0  ...             ocr            PF08684.13        NaN  NaN  NaN  NaN  \\\n",
       "1  ...             NaN                   NaN        NaN  NaN  NaN  NaN   \n",
       "2  ...             NaN                   NaN        NaN  NaN  NaN  NaN   \n",
       "3  ...             NaN                   NaN   2.7.11.1  NaN  NaN  NaN   \n",
       "4  ...  RNA_pol,RPOL_N  PF00940.22,PF14700.9    2.7.7.6    1    1  NaN   \n",
       "\n",
       "  cdd_name tigr tigrfam_name conj_pl_domains  \n",
       "0      NaN  NaN          NaN             NaN  \n",
       "1      NaN  NaN          NaN             NaN  \n",
       "2      NaN  NaN          NaN             NaN  \n",
       "3      NaN  NaN          NaN             NaN  \n",
       "4      NaN  NaN          NaN             NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "17e78015",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([   0,    3,    5,    8,   14,   23,   52,   89,  124,  141,\n",
       "       ...\n",
       "       3038, 3045, 3058, 3063, 3082, 3089, 3103, 3110, 3138, 3167],\n",
       "      dtype='int64', length=158)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_data[gff_data['cluster_num'].isin(clusters_ocr)].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b2cd26b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json(json_like: str) -> dict:\n",
    "    \"\"\"\n",
    "    read json string, returns it as a (dict) object\n",
    "    \"\"\"\n",
    "    # based on file structure:\n",
    "    json_like = json_like.replace(\" '\", ' \"')\n",
    "    json_like = json_like.replace(\"{'\", '{\"')\n",
    "    json_like = json_like.replace(\"':\", '\":')\n",
    "\n",
    "    js_d = json.loads(json_like)\n",
    "    return js_d\n",
    "\n",
    "\n",
    "def define_clu_main_product(products: dict) -> str:\n",
    "    \"\"\"\n",
    "\n",
    "    :param functions: (dict) contains of all functions of cluster\n",
    "    :return: main_function (str) name of cluster's main functional product\n",
    "    \"\"\"\n",
    "    main_function = None\n",
    "    max_count = 0\n",
    "    for product, count in products.items():\n",
    "        if product != 'hypothetical protein' and count > max_count:\n",
    "            main_function = product\n",
    "            max_count = count\n",
    "    if not main_function and ('hypothetical protein' in products.keys()):\n",
    "        main_function = 'hypothetical protein'\n",
    "    return main_function\n",
    "\n",
    "\n",
    "def extract_cluster_functions_wide(wide_path: str) -> dict:\n",
    "    \"\"\"\n",
    "    extract information about main function of each cluster (dict) {cluster_parent: (number of cluster, function)}\n",
    "    :param wide_path:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    cluster_to_function = {}\n",
    "    with open(wide_path, 'r') as tsv_file:\n",
    "        for ind, line in enumerate(tsv_file):\n",
    "            cluster = line.strip().split('\\t')\n",
    "            clu_parent = cluster[0]\n",
    "            clu_functs = cluster[2]\n",
    "            clu_funct_js = read_json(clu_functs)\n",
    "            clu_main_func = define_clu_main_product(clu_funct_js)\n",
    "            cluster_to_function[clu_parent] = (ind + 1, clu_main_func)\n",
    "    return cluster_to_function\n",
    "\n",
    "\n",
    "def extract_cluster_children_long(long_path: str) -> dict:\n",
    "    \"\"\"\n",
    "    extract information about children in clusters, returns dict {child: parent}\n",
    "    :param long_path: path to long file with results of upstream gene products clusters\n",
    "    :return: (dict) {child: parent} in cluster\n",
    "    \"\"\"\n",
    "    childs_to_parent = {}\n",
    "\n",
    "    with open(long_path, 'r') as tsv_file:\n",
    "        for line in tsv_file:\n",
    "            cluster = line.strip().split('\\t')\n",
    "            clu_parent = cluster[0]\n",
    "            clu_child = cluster[1]\n",
    "            childs_to_parent[clu_child] = clu_parent\n",
    "\n",
    "    return childs_to_parent\n",
    "\n",
    "\n",
    "def read_upstream_gff(gff_path: str) -> dict:\n",
    "    upstreams = OrderedDict()\n",
    "    with open(gff_path, 'r') as gff_file:\n",
    "        for line in gff_file:\n",
    "            if not line.startswith('#'):\n",
    "                orf_info = line.strip().split('\\t')\n",
    "                chrom = orf_info[0]\n",
    "                if chrom not in upstreams.keys():\n",
    "                    upstreams[chrom] = [orf_info]\n",
    "                else:\n",
    "                    upstreams[chrom].append(orf_info)\n",
    "    return upstreams\n",
    "\n",
    "\n",
    "def add_cluster_information(upstreams: dict, clu_children: dict, clu_products: dict) -> dict:\n",
    "    \"\"\"\n",
    "    add information about cluster for each cds in gff dict object:\n",
    "    {genome: [[protein_1 gff], [protein_2 gff]]} -> add_cluster_information ->\n",
    "    -> {genome: [[protein_1 gff, (num_1, function_1)],\n",
    "                 [protein_2 gff, (num_2, function_2)]]\n",
    "        }\n",
    "    \"\"\"\n",
    "    upstreams_mod = upstreams.copy()\n",
    "\n",
    "    for genome, cdss in upstreams.items():\n",
    "        for ind, cds in enumerate(cdss):\n",
    "            attribute = cds[8]\n",
    "            attribute_list = attribute.split(';')\n",
    "            cds_id = attribute_list[0].split('=')[1]\n",
    "            parent = clu_children[cds_id]\n",
    "            function = clu_products[parent]\n",
    "            upstreams_mod[genome][ind].append(function)\n",
    "    return upstreams_mod\n",
    "\n",
    "\n",
    "def _find_numclu_one(clu_products: dict, keyword: str) -> set:\n",
    "    \"\"\"\n",
    "    :param clu_products (tuple): (num_clu, main_product)\n",
    "    :param keyword (str): product name, needed to find and exclude\n",
    "    :return: (set) of cluster numbers to exclude\n",
    "    \"\"\"\n",
    "    tabu_clu_num = set()\n",
    "    for num, func in clu_products.values():\n",
    "        if re.findall(keyword, func, re.IGNORECASE):\n",
    "            tabu_clu_num.add(num)\n",
    "    return tabu_clu_num\n",
    "\n",
    "\n",
    "def find_numclu_to_filter(clu_products: dict, keywords: set) -> set:\n",
    "    \"\"\"\n",
    "    uses `_find_numclu_one` for set of keywords\n",
    "    :param clu_products: clu_products (tuple): (num_clu, main_product)\n",
    "    :param keywords (set): of product names, needed to find and exclude\n",
    "    :return: (set) of cluster numbers to exclude\n",
    "    \"\"\"\n",
    "    tabu_clu_nums = set()\n",
    "    for keyword in keywords:\n",
    "        tabu_clu_nums = tabu_clu_nums | _find_numclu_one(clu_products, keyword)\n",
    "    return tabu_clu_nums\n",
    "\n",
    "\n",
    "def filter_with_conditions(upstream_mod):\n",
    "\n",
    "    if sum(upstream_mod['strand'] == '-') == len(upstream_mod):  # potential error place\n",
    "        select_indices = list(np.where(upstream_mod['cluster_in_tabu'])[0])\n",
    "        target_index = select_indices[0]\n",
    "    elif sum(upstream_mod['strand'] == '+') == len(upstream_mod):\n",
    "        select_indices = list(np.where(upstream_mod['cluster_in_tabu'])[0])\n",
    "        target_index = select_indices[-1]\n",
    "    else:\n",
    "        chormosome = upstream_mod.chromosome.unique()[0]\n",
    "        raise ValueError(f'Non-unique strand for {chormosome}')\n",
    "\n",
    "    start = upstream_mod.iloc[target_index]['start']\n",
    "    end = upstream_mod.iloc[target_index]['end']\n",
    "\n",
    "    if sum(upstream_mod['strand'] == '-') == len(upstream_mod):\n",
    "        condition = (upstream_mod['end'] <= end) | (upstream_mod['start'] - 5000 >= start)\n",
    "    else:\n",
    "        condition = (upstream_mod['end'] >= end) | (upstream_mod['start'] + 5000 <= start)\n",
    "\n",
    "    upstream_mod = upstream_mod[condition]\n",
    "    return upstream_mod\n",
    "\n",
    "\n",
    "def filter_within_one_genome(upstream: list, tabu_clusters: set) -> list:\n",
    "    \"\"\"\n",
    "    finds tabu genes in one genome and returns the resulting list without it's upstream region\n",
    "    :param upstream (list): part of upstream.gff for one region:\n",
    "    {genome (list): [[protein_1 gff, (num_1, function_1)],\n",
    "                 [protein_2 gff, (num_2, function_2)]]\n",
    "        }\n",
    "    :param tabu_clu_nums (set): num of tabu\n",
    "    :return: upstream_mod (list) with removed tabu genes and their upstream\n",
    "    \"\"\"\n",
    "    colnames_ = ['chromosome', 'annotator', 'type', 'start', 'end',\n",
    "                 'something', 'strand', 'somthing_else', 'attribute', 'cluster']\n",
    "\n",
    "    upstream_df = pd.DataFrame(upstream, columns=colnames_)  # list to dataframe\n",
    "    # numbers: str to int\n",
    "    upstream_df['start'] = list(map(lambda x: int(x), upstream_df['start']))\n",
    "    upstream_df['end'] = list(map(lambda x: int(x), upstream_df['end']))\n",
    "    # column: cluster number:\n",
    "    upstream_df['cluster_num'] = list(map(lambda x: x[0], upstream_df['cluster']))\n",
    "    # column: is tabu cluster?\n",
    "    upstream_df['cluster_in_tabu'] = list(map(lambda x: x in tabu_clusters, upstream_df['cluster_num']))\n",
    "\n",
    "    if sum(upstream_df['cluster_in_tabu']) > 0:\n",
    "        upstream_df = filter_with_conditions(upstream_df)\n",
    "    upstream_mod = upstream_df.iloc[:, :10].values.tolist()\n",
    "    return upstream_mod\n",
    "\n",
    "\n",
    "def make_gff_with_clusters(path_long: str, path_wide: str, path_upstream: str):\n",
    "    children_to_clu = extract_cluster_children_long(path_long)\n",
    "    parent_to_func = extract_cluster_functions_wide(path_wide)\n",
    "    upstreams = read_upstream_gff(path_upstream)\n",
    "    ups_with_clu_info = add_cluster_information(upstreams=upstreams,\n",
    "                                                clu_children=children_to_clu,\n",
    "                                                clu_products=parent_to_func)\n",
    "    upstreams_mod = []\n",
    "    for upstream in ups_with_clu_info.values():\n",
    "        upstreams_mod.append(upstream)\n",
    "    return upstreams_mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "45754d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p_wide = 'results/upstream_proteins_clu_wide_seq_sorted_prokka_domains_pi_length_host.tsv'\n",
    "# p_l = 'results/upstream_proteins_clu_long_prokka.tsv'\n",
    "# gff_file \n",
    "path_wide_ = os.path.join(res_dir, 'upstream_proteins_clu_wide_seq_sorted_prokka.tsv')\n",
    "path_long_ = os.path.join(res_dir, 'upstream_proteins_clu_long_prokka.tsv')\n",
    "upstreams = make_gff_with_clusters(path_long=path_long_, path_wide=path_wide_, path_upstream=gff_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8cceec01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AY264776.1</td>\n",
       "      <td>Prodigal:2.6</td>\n",
       "      <td>CDS</td>\n",
       "      <td>901</td>\n",
       "      <td>1254</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=GCA_002600025.1_00001;file_id=upstreams_wit...</td>\n",
       "      <td>(3, Overcome classical restriction gp0.3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AY264776.1</td>\n",
       "      <td>Prodigal:2.6</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1254</td>\n",
       "      <td>1409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=GCA_002600025.1_00002;file_id=upstreams_wit...</td>\n",
       "      <td>(13, Gene 0.4 protein)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY264776.1</td>\n",
       "      <td>Prodigal:2.6</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1557</td>\n",
       "      <td>4208</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=GCA_002600025.1_00003;Name=1;file_id=upstre...</td>\n",
       "      <td>(1, T7 RNA polymerase)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0             1    2     3     4  5  6  7   \n",
       "0  AY264776.1  Prodigal:2.6  CDS   901  1254  .  +  0  \\\n",
       "1  AY264776.1  Prodigal:2.6  CDS  1254  1409  .  +  0   \n",
       "2  AY264776.1  Prodigal:2.6  CDS  1557  4208  .  +  0   \n",
       "\n",
       "                                                   8   \n",
       "0  ID=GCA_002600025.1_00001;file_id=upstreams_wit...  \\\n",
       "1  ID=GCA_002600025.1_00002;file_id=upstreams_wit...   \n",
       "2  ID=GCA_002600025.1_00003;Name=1;file_id=upstre...   \n",
       "\n",
       "                                           9  \n",
       "0  (3, Overcome classical restriction gp0.3)  \n",
       "1                     (13, Gene 0.4 protein)  \n",
       "2                     (1, T7 RNA polymerase)  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(upstreams[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "cc7a5e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames_ = ['chromosome', 'annotator', 'type', 'start', 'end',\n",
    "             'something', 'strand', 'somthing_else', 'attribute', 'cluster']\n",
    "\n",
    "\n",
    "def find_right_neighbour(upstream, gene, index):\n",
    "    strand = gene[6]\n",
    "    if strand == '+':\n",
    "        if index != len(upstream) - 1:\n",
    "            right_neighbour = upstream[index + 1]\n",
    "            gene_end = int(gene[4])\n",
    "            right_neighbour_start = int(right_neighbour[3])\n",
    "            if gene_end + 5000 > right_neighbour_start:\n",
    "                return right_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "        else:\n",
    "            right_neighbour = upstream[0]\n",
    "            gene_start = int(gene[3])\n",
    "            right_neighbour_end = int(right_neighbour[4])\n",
    "            if gene_start - 5000 > right_neighbour_end: \n",
    "                return right_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "    else:\n",
    "        if index != 0:\n",
    "            right_neighbour = upstream[index - 1]\n",
    "            gene_start = int(gene[3])\n",
    "            right_neighbour_end = int(right_neighbour[4])\n",
    "            if gene_start - 5000 < right_neighbour_end:\n",
    "                return right_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "        else:\n",
    "            right_neighbour = upstream[-1]\n",
    "            gene_end = int(gene[4])\n",
    "            right_neighbour_start = int(right_neighbour[3])\n",
    "            if gene_end + 5000 < right_neighbour_start:\n",
    "                 return right_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "        \n",
    "\n",
    "right_neighbours = defaultdict(list)\n",
    "\n",
    "for upstream in upstreams:\n",
    "    for i, gene in enumerate(upstream):\n",
    "        cluster = gene[9][0]\n",
    "        right_neighbours[cluster].append(find_right_neighbour(upstream, gene, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4a79fb45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_left_neighbour(upstream, gene, index):\n",
    "    strand = gene[6]\n",
    "    if strand == '-':\n",
    "        if index != len(upstream) - 1:\n",
    "            left_neighbour = upstream[index + 1]\n",
    "            gene_end = int(gene[4])\n",
    "            left_neighbour_start = int(left_neighbour[3])\n",
    "            if gene_end + 5000 > left_neighbour_start:\n",
    "                return left_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "        else:\n",
    "            left_neighbour = upstream[0]\n",
    "            gene_start = int(gene[3])\n",
    "            left_neighbour_end = int(left_neighbour[4])\n",
    "            if gene_start - 5000 > left_neighbour_end: \n",
    "                return left_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "    else:\n",
    "        if index != 0:\n",
    "            left_neighbour = upstream[index - 1]\n",
    "            gene_start = int(gene[3])\n",
    "            left_neighbour_end = int(left_neighbour[4])\n",
    "            if gene_start - 5000 < left_neighbour_end:\n",
    "                return left_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "        else:\n",
    "            left_neighbour = upstream[-1]\n",
    "            gene_end = int(gene[4])\n",
    "            left_neighbour_start = int(left_neighbour[3])\n",
    "            if gene_end + 5000 < left_neighbour_start:\n",
    "                 return left_neighbour[9][0]\n",
    "            else:\n",
    "                return None\n",
    "        \n",
    "\n",
    "left_neighbours = defaultdict(list)\n",
    "\n",
    "for upstream in upstreams:\n",
    "    for i, gene in enumerate(upstream):\n",
    "        cluster = gene[9][0]\n",
    "        left_neighbours[cluster].append(find_left_neighbour(upstream, gene, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2b7212b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_neighbours(neighbours_dict: dict, clu_nums: str) -> set:\n",
    "    right_neighbours = set()\n",
    "    for clu in clu_nums:\n",
    "        for neighbour in neighbours_dict[clu]:  # 0-based\n",
    "            right_neighbours.add(neighbour)  # 1-based\n",
    "    return right_neighbours\n",
    "\n",
    "\n",
    "ocr_neighbours = extract_neighbours(right_neighbours, clusters_ocr)\n",
    "samase_neighbours = extract_neighbours(right_neighbours, clusters_samase)\n",
    "ard_neighbours = extract_neighbours(right_neighbours, clusters_ard)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "011dfa69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ocr_neighbours)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "065f1ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_common_neighbours_one_clu(neigbours_dict: dict, target_neighbours: set, clu: int) -> set:\n",
    "    neighbours_clu = set(neigbours_dict[clu])\n",
    "    common_ns = target_neighbours & neighbours_clu\n",
    "    return common_ns\n",
    "\n",
    "\n",
    "def get_common_neighbours(neigbours_dict: dict, target_neighbours: set, n: int) -> dict:\n",
    "    common_neighbours = {}\n",
    "    for clu in range(1, n):\n",
    "        common = get_common_neighbours_one_clu(neigbours_dict, target_neighbours, clu)\n",
    "        common_neighbours[clu] = common\n",
    "    return common_neighbours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d96b97e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common right neighbours for Ocr and SAMase: {32, 19, 21, 6}\n",
      "Common right neighbours for Ocr and ArdA/ArdcN: {139}\n",
      "Common right neighbours for SAMase and ArdA/ArdcN: set()\n"
     ]
    }
   ],
   "source": [
    "n = len(df_agg)+2\n",
    "\n",
    "ocr_neighbours = extract_neighbours(right_neighbours, clusters_ocr)\n",
    "samase_neighbours = extract_neighbours(right_neighbours, clusters_samase)\n",
    "ard_neighbours = extract_neighbours(right_neighbours, clusters_ard)\n",
    "print(f'Common right neighbours for Ocr and SAMase: {ocr_neighbours & samase_neighbours}')\n",
    "print(f'Common right neighbours for Ocr and ArdA/ArdcN: {ocr_neighbours & ard_neighbours}')\n",
    "print(f'Common right neighbours for SAMase and ArdA/ArdcN: {samase_neighbours & ard_neighbours}')\n",
    "\n",
    "res_table['common_ocr_right_neighbour'] = res_table['cluster_num'].apply(lambda x: \n",
    "                                                                         get_common_neighbours_one_clu(right_neighbours, \n",
    "                                                                                                       ocr_neighbours, x))\n",
    "res_table['common_samase_right_neighbour'] = res_table['cluster_num'].apply(lambda x: \n",
    "                                                                         get_common_neighbours_one_clu(right_neighbours, \n",
    "                                                                                                       samase_neighbours, x))\n",
    "res_table['common_ard_right_neighbour'] = res_table['cluster_num'].apply(lambda x: \n",
    "                                                                         get_common_neighbours_one_clu(right_neighbours, \n",
    "                                                                                                       ard_neighbours, x))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "dc770afd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common left neighbours for Ocr and SAMase: {41, 18, None}\n",
      "Common left neighbours for Ocr and ArdA/ArdcN: {None, 14}\n",
      "Common left neighbours for SAMase and ArdA/ArdcN: {None}\n"
     ]
    }
   ],
   "source": [
    "ocr_neighbours = extract_neighbours(left_neighbours, clusters_ocr)\n",
    "samase_neighbours = extract_neighbours(left_neighbours, clusters_samase)\n",
    "ard_neighbours = extract_neighbours(left_neighbours, clusters_ard)\n",
    "\n",
    "print(f'Common left neighbours for Ocr and SAMase: {ocr_neighbours & samase_neighbours}')\n",
    "print(f'Common left neighbours for Ocr and ArdA/ArdcN: {ocr_neighbours & ard_neighbours}')\n",
    "print(f'Common left neighbours for SAMase and ArdA/ArdcN: {samase_neighbours & ard_neighbours}')\n",
    "\n",
    "res_table['common_ocr_left_neighbour'] = res_table['cluster_num'].apply(lambda x: \n",
    "                                                                         get_common_neighbours_one_clu(left_neighbours, \n",
    "                                                                                                       ocr_neighbours, x))\n",
    "res_table['common_samase_left_neighbour'] = res_table['cluster_num'].apply(lambda x: \n",
    "                                                                         get_common_neighbours_one_clu(left_neighbours, \n",
    "                                                                                                       samase_neighbours, x))\n",
    "res_table['common_ard_left_neighbour'] = res_table['cluster_num'].apply(lambda x: \n",
    "                                                                         get_common_neighbours_one_clu(left_neighbours, \n",
    "                                                                                                       ard_neighbours, x))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d6a470ef",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clu_parent</th>\n",
       "      <th>count</th>\n",
       "      <th>products</th>\n",
       "      <th>sequence</th>\n",
       "      <th>phrOG</th>\n",
       "      <th>phrOG_annotation</th>\n",
       "      <th>phrOG_category</th>\n",
       "      <th>pfam_annotation</th>\n",
       "      <th>pfam</th>\n",
       "      <th>conj_pl_domains</th>\n",
       "      <th>...</th>\n",
       "      <th>LQLenths</th>\n",
       "      <th>UQLenths</th>\n",
       "      <th>Ecoli_host</th>\n",
       "      <th>Enterobacter_host</th>\n",
       "      <th>common_ocr_right_neighbour</th>\n",
       "      <th>common_samase_right_neighbour</th>\n",
       "      <th>common_ard_right_neighbour</th>\n",
       "      <th>common_ocr_left_neighbour</th>\n",
       "      <th>common_samase_left_neighbour</th>\n",
       "      <th>common_ard_left_neighbour</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_001470995.1_00023</td>\n",
       "      <td>354</td>\n",
       "      <td>{T7 RNA polymerase: 354}</td>\n",
       "      <td>MEYNPLTELASLYGEDLAAEQLRLEAEAYSLGEKRFMEAMEFKAET...</td>\n",
       "      <td>phrog_20599,phrog_414</td>\n",
       "      <td>RNA polymerase</td>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>RNA_pol,RPOL_N</td>\n",
       "      <td>PF00940.22,PF14700.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>883.0</td>\n",
       "      <td>906.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{14}</td>\n",
       "      <td>{3}</td>\n",
       "      <td>{14}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_025085155.1_00014</td>\n",
       "      <td>190</td>\n",
       "      <td>{Protein kinase 0.7: 183, hypothetical protein...</td>\n",
       "      <td>MYAHFQSLLKAIRELPINELDKRQPMLVDLLAQIVNHETQDGAITA...</td>\n",
       "      <td>phrog_18305,phrog_2828,phrog_15752</td>\n",
       "      <td>kinase,serine-threonine kinase,protein kinase;...</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>337.0</td>\n",
       "      <td>369.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>{1}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{19, 14}</td>\n",
       "      <td>{3}</td>\n",
       "      <td>{14}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_001041815.1_00006</td>\n",
       "      <td>150</td>\n",
       "      <td>{Overcome classical restriction gp0.3: 102, hy...</td>\n",
       "      <td>MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...</td>\n",
       "      <td>phrog_2568</td>\n",
       "      <td>ocr-like anti-restriction</td>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>ocr</td>\n",
       "      <td>PF08684.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>114.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>{1, 2, 5, 6, 8, 10, 12, 13, 16, 19, 21, 32, 56...</td>\n",
       "      <td>{32, 19, 21, 6}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{5, 647, 11, 14, 18, 22, 534, 29, 681, 41, 45,...</td>\n",
       "      <td>{41, 18, None}</td>\n",
       "      <td>{None, 14}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_023716895.1_00035</td>\n",
       "      <td>149</td>\n",
       "      <td>{hypothetical protein: 149}</td>\n",
       "      <td>METSCRLPHGAGRSCTGMDGDAMMFTAFIVFMYLLIVLYFLKDFRK...</td>\n",
       "      <td>phrog_2385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>{1, 2, 13}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{18, 11, 5}</td>\n",
       "      <td>{18, 77}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_009672325.1_00007</td>\n",
       "      <td>144</td>\n",
       "      <td>{hypothetical protein: 144}</td>\n",
       "      <td>MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...</td>\n",
       "      <td>phrog_2226</td>\n",
       "      <td>SAM-dependent methyltransferase</td>\n",
       "      <td>other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>152.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>{32, 19, 21, 6}</td>\n",
       "      <td>{512, 3, 4, 645, 6, 11, 14, 18, 19, 21, 24, 32...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{41, 18, None}</td>\n",
       "      <td>{3, 230, 41, 265, 141, 18, None, 61}</td>\n",
       "      <td>{None}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              clu_parent  count   \n",
       "0  GCF_001470995.1_00023    354  \\\n",
       "1  GCA_025085155.1_00014    190   \n",
       "2  GCF_001041815.1_00006    150   \n",
       "3  GCA_023716895.1_00035    149   \n",
       "4  GCA_009672325.1_00007    144   \n",
       "\n",
       "                                            products   \n",
       "0                           {T7 RNA polymerase: 354}  \\\n",
       "1  {Protein kinase 0.7: 183, hypothetical protein...   \n",
       "2  {Overcome classical restriction gp0.3: 102, hy...   \n",
       "3                        {hypothetical protein: 149}   \n",
       "4                        {hypothetical protein: 144}   \n",
       "\n",
       "                                            sequence   \n",
       "0  MEYNPLTELASLYGEDLAAEQLRLEAEAYSLGEKRFMEAMEFKAET...  \\\n",
       "1  MYAHFQSLLKAIRELPINELDKRQPMLVDLLAQIVNHETQDGAITA...   \n",
       "2  MSNVTYHSLYLSAKQGLINVIRDNDLRDYDEACDYVHEVADIYVPH...   \n",
       "3  METSCRLPHGAGRSCTGMDGDAMMFTAFIVFMYLLIVLYFLKDFRK...   \n",
       "4  MIYTNEPANIFYMYVSAARADRPDVVNLARHNQLKRDIMNGLGMYG...   \n",
       "\n",
       "                                phrOG   \n",
       "0               phrog_20599,phrog_414  \\\n",
       "1  phrog_18305,phrog_2828,phrog_15752   \n",
       "2                          phrog_2568   \n",
       "3                          phrog_2385   \n",
       "4                          phrog_2226   \n",
       "\n",
       "                                    phrOG_annotation   \n",
       "0                                     RNA polymerase  \\\n",
       "1  kinase,serine-threonine kinase,protein kinase;...   \n",
       "2                          ocr-like anti-restriction   \n",
       "3                                                NaN   \n",
       "4                    SAM-dependent methyltransferase   \n",
       "\n",
       "                                      phrOG_category pfam_annotation   \n",
       "0                 DNA, RNA and nucleotide metabolism  RNA_pol,RPOL_N  \\\n",
       "1                                              other             NaN   \n",
       "2  moron, auxiliary metabolic gene and host takeover             ocr   \n",
       "3                                   unknown function             NaN   \n",
       "4                                              other             NaN   \n",
       "\n",
       "                   pfam conj_pl_domains  ... LQLenths UQLenths Ecoli_host   \n",
       "0  PF00940.22,PF14700.9             NaN  ...    883.0    906.0       True  \\\n",
       "1                   NaN             NaN  ...    337.0    369.0       True   \n",
       "2            PF08684.13             NaN  ...    114.0    121.0       True   \n",
       "3                   NaN             NaN  ...     65.0     65.0       True   \n",
       "4                   NaN             NaN  ...    152.0    155.0       True   \n",
       "\n",
       "  Enterobacter_host                         common_ocr_right_neighbour   \n",
       "0              True                                                 {}  \\\n",
       "1              True                                                {1}   \n",
       "2              True  {1, 2, 5, 6, 8, 10, 12, 13, 16, 19, 21, 32, 56...   \n",
       "3              True                                         {1, 2, 13}   \n",
       "4              True                                    {32, 19, 21, 6}   \n",
       "\n",
       "                       common_samase_right_neighbour   \n",
       "0                                                 {}  \\\n",
       "1                                                 {}   \n",
       "2                                    {32, 19, 21, 6}   \n",
       "3                                                 {}   \n",
       "4  {512, 3, 4, 645, 6, 11, 14, 18, 19, 21, 24, 32...   \n",
       "\n",
       "   common_ard_right_neighbour   \n",
       "0                          {}  \\\n",
       "1                          {}   \n",
       "2                          {}   \n",
       "3                          {}   \n",
       "4                          {}   \n",
       "\n",
       "                           common_ocr_left_neighbour   \n",
       "0                                               {14}  \\\n",
       "1                                           {19, 14}   \n",
       "2  {5, 647, 11, 14, 18, 22, 534, 29, 681, 41, 45,...   \n",
       "3                                        {18, 11, 5}   \n",
       "4                                     {41, 18, None}   \n",
       "\n",
       "           common_samase_left_neighbour  common_ard_left_neighbour  \n",
       "0                                   {3}                       {14}  \n",
       "1                                   {3}                       {14}  \n",
       "2                        {41, 18, None}                 {None, 14}  \n",
       "3                              {18, 77}                         {}  \n",
       "4  {3, 230, 41, 265, 141, 18, None, 61}                     {None}  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b324db5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_table.to_csv(output_file, sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
